{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2685f3a6-cf6b-4c7d-bac3-37ca3d5157d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently Loaded Modulefiles:\n",
      " 1) \u001b[46mpowerPlant/core\u001b[0m    5) \u001b[46mSlurm/21.08.8-2\u001b[0m      9) bowtie2/2.3.4.3  \n",
      " 2) texlive/20230914   6) perlbrew/0.76       10) minimap2/2.22    \n",
      " 3) pandoc/1.19.2      7) perl/5.36.0         11) cd-hit/4.6.1     \n",
      " 4) git/2.21.0         8) slurm-utils/latest  \n",
      "\n",
      "Key:\n",
      "\u001b[46msticky\u001b[0m  \n"
     ]
    }
   ],
   "source": [
    "# Alignments for various species to reference genomes\n",
    "\n",
    "# loading modules\n",
    "module load bowtie2\n",
    "module load minimap2 \n",
    "module load cd-hit\n",
    "\n",
    "# checking active \n",
    "# module list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "3eec4422-fd71-4360-90d1-1fe9b4aaddfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# setting up variables \n",
    "# paths\n",
    "path_raw=/workspace/cfngle/raw-data/\n",
    "path_results=/workspace/cfngle/results-data/\n",
    "\n",
    "# filenames\n",
    "# 200 bp sequences\n",
    "seq_200bp=(\"AC_CpG_200bp.fasta\" \"AS_CpG_200bp.fasta\" \"EH_CpG_200bp.fasta\")\n",
    "\n",
    "seq_100bp=(\"AC_CpG_100bp.fasta\" \"AS_CpG_100bp.fasta\" \"EH_CpG_100bp.fasta\")\n",
    "\n",
    ": << 'COMMENT'\n",
    "n=${#seq_200bp[@]}\n",
    "for i in ${seq_200bp[@]}; do\n",
    "    echo $i\n",
    "    echo $n\n",
    "    echo \"${seq_200bp[n]}\"\n",
    "    n=$((n-1))\n",
    "done\n",
    "COMMENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e6b03971-a762-4410-9d0c-7ea102ede3a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================================================================\n",
      "Program: CD-HIT, V4.6 (+OpenMP), Jun 26 2013, 16:45:20\n",
      "Command: cd-hit-est-2d -i\n",
      "         /workspace/cfngle/results-data/AC_CpG_200bp.fasta -i2\n",
      "         /workspace/cfngle/results-data/AS_CpG_200bp.fasta -o\n",
      "         /workspace/cfngle/results-data/cd-hit/AC_AS_cd-hit_85.fasta\n",
      "         -p 1 -c 0.85 -n 6 -T 8 -M 8000\n",
      "\n",
      "Started: Fri Feb  9 13:09:47 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq in db1: 13755\n",
      "total seq in db2: 24535\n",
      "longest and shortest : 635 and 200\n",
      "Total letters: 3867239\n",
      "Sequences have been sorted\n",
      "longest and shortest : 579 and 200\n",
      "Total letters: 7326339\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 14M\n",
      "Buffer          : 8 X 12M = 98M\n",
      "Table           : 2 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 114M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 4000000\n",
      "Max number of word counting entries: 985739687\n",
      "\n",
      "..........    10000  finished\n",
      "..........        0  compared         13  clusters\n",
      "\n",
      "24535 compared\t13 clustered\n",
      "writing non-redundant sequences from db2\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 568.32\n",
      "================================================================\n",
      "Program: CD-HIT, V4.6 (+OpenMP), Jun 26 2013, 16:45:20\n",
      "Command: cd-hit-est-2d -i\n",
      "         /workspace/cfngle/results-data/AC_CpG_200bp.fasta -i2\n",
      "         /workspace/cfngle/results-data/EH_CpG_200bp.fasta -o\n",
      "         /workspace/cfngle/results-data/cd-hit/AC_EH_cd-hit_85.fasta\n",
      "         -p 1 -c 0.85 -n 6 -T 8 -M 8000\n",
      "\n",
      "Started: Fri Feb  9 13:11:04 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq in db1: 13755\n",
      "total seq in db2: 18337\n",
      "longest and shortest : 635 and 200\n",
      "Total letters: 3867239\n",
      "Sequences have been sorted\n",
      "longest and shortest : 941 and 200\n",
      "Total letters: 6258566\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 12M\n",
      "Buffer          : 8 X 12M = 99M\n",
      "Table           : 2 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 112M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 4000000\n",
      "Max number of word counting entries: 985900866\n",
      "\n",
      "..........    10000  finished\n",
      "..........        0  compared        128  clusters\n",
      "\n",
      "18337 compared\t128 clustered\n",
      "writing non-redundant sequences from db2\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 302.58\n",
      "================================================================\n",
      "Program: CD-HIT, V4.6 (+OpenMP), Jun 26 2013, 16:45:20\n",
      "Command: cd-hit-est-2d -i\n",
      "         /workspace/cfngle/results-data/AS_CpG_200bp.fasta -i2\n",
      "         /workspace/cfngle/results-data/EH_CpG_200bp.fasta -o\n",
      "         /workspace/cfngle/results-data/cd-hit/AS_EH_cd-hit_85.fasta\n",
      "         -p 1 -c 0.85 -n 6 -T 8 -M 8000\n",
      "\n",
      "Started: Fri Feb  9 13:11:47 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq in db1: 24535\n",
      "total seq in db2: 18337\n",
      "longest and shortest : 579 and 200\n",
      "Total letters: 7326339\n",
      "Sequences have been sorted\n",
      "longest and shortest : 941 and 200\n",
      "Total letters: 6258566\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 16M\n",
      "Buffer          : 8 X 12M = 100M\n",
      "Table           : 2 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 117M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 4000000\n",
      "Max number of word counting entries: 985269016\n",
      "\n",
      "..........    10000  finished\n",
      "..........    20000  finished\n",
      "..........        0  compared         69  clusters\n",
      "\n",
      "18337 compared\t69 clustered\n",
      "writing non-redundant sequences from db2\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 1146.57\n",
      "================================================================\n",
      "Program: CD-HIT, V4.6 (+OpenMP), Jun 26 2013, 16:45:20\n",
      "Command: cd-hit-est-2d -i\n",
      "         /workspace/cfngle/results-data/AS_CpG_200bp.fasta -i2\n",
      "         /workspace/cfngle/results-data/AC_CpG_200bp.fasta -o\n",
      "         /workspace/cfngle/results-data/cd-hit/AS_AC_cd-hit_85.fasta\n",
      "         -p 1 -c 0.85 -n 6 -T 8 -M 8000\n",
      "\n",
      "Started: Fri Feb  9 13:14:23 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq in db1: 24535\n",
      "total seq in db2: 13755\n",
      "longest and shortest : 579 and 200\n",
      "Total letters: 7326339\n",
      "Sequences have been sorted\n",
      "longest and shortest : 635 and 200\n",
      "Total letters: 3867239\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 13M\n",
      "Buffer          : 8 X 12M = 99M\n",
      "Table           : 2 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 114M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 4000000\n",
      "Max number of word counting entries: 985713868\n",
      "\n",
      "..........    10000  finished\n",
      "..........    20000  finished\n",
      "..........        0  compared         60  clusters\n",
      "\n",
      "13755 compared\t60 clustered\n",
      "writing non-redundant sequences from db2\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 1368.23\n",
      "================================================================\n",
      "Program: CD-HIT, V4.6 (+OpenMP), Jun 26 2013, 16:45:20\n",
      "Command: cd-hit-est-2d -i\n",
      "         /workspace/cfngle/results-data/EH_CpG_200bp.fasta -i2\n",
      "         /workspace/cfngle/results-data/AC_CpG_200bp.fasta -o\n",
      "         /workspace/cfngle/results-data/cd-hit/EH_AC_cd-hit_85.fasta\n",
      "         -p 1 -c 0.85 -n 6 -T 8 -M 8000\n",
      "\n",
      "Started: Fri Feb  9 13:17:35 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq in db1: 18337\n",
      "total seq in db2: 13755\n",
      "longest and shortest : 941 and 200\n",
      "Total letters: 6258566\n",
      "Sequences have been sorted\n",
      "longest and shortest : 635 and 200\n",
      "Total letters: 3867239\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 12M\n",
      "Buffer          : 8 X 12M = 99M\n",
      "Table           : 2 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 112M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 4000000\n",
      "Max number of word counting entries: 985961999\n",
      "\n",
      "..........    10000  finished\n",
      "..........        0  compared        828  clusters\n",
      "\n",
      "13755 compared\t828 clustered\n",
      "writing non-redundant sequences from db2\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 1651.21\n",
      "================================================================\n",
      "Program: CD-HIT, V4.6 (+OpenMP), Jun 26 2013, 16:45:20\n",
      "Command: cd-hit-est-2d -i\n",
      "         /workspace/cfngle/results-data/EH_CpG_200bp.fasta -i2\n",
      "         /workspace/cfngle/results-data/AS_CpG_200bp.fasta -o\n",
      "         /workspace/cfngle/results-data/cd-hit/EH_AS_cd-hit_85.fasta\n",
      "         -p 1 -c 0.85 -n 6 -T 8 -M 8000\n",
      "\n",
      "Started: Fri Feb  9 13:21:22 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq in db1: 18337\n",
      "total seq in db2: 24535\n",
      "longest and shortest : 941 and 200\n",
      "Total letters: 6258566\n",
      "Sequences have been sorted\n",
      "longest and shortest : 579 and 200\n",
      "Total letters: 7326339\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 17M\n",
      "Buffer          : 8 X 12M = 99M\n",
      "Table           : 2 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 117M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 4000000\n",
      "Max number of word counting entries: 985355968\n",
      "\n",
      "..........    10000  finished\n",
      "..........        0  compared         81  clusters\n",
      "\n",
      "24535 compared\t81 clustered\n",
      "writing non-redundant sequences from db2\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 2545.62\n",
      "bash: \n",
      "for i in AC_CpG_200bp.fasta; do\n",
      "    cd-hit-est-2d -i /workspace/cfngle/results-data/EH_CpG_200bp.fasta -i2 /workspace/cfngle/results-data/AC -o /workspace/cfngle/results-data/cd-hit/EH_CpG_200bp.fastacd-hit_AC.fasta -c 0.9 -p 1\n",
      "done\n",
      ": No such file or directory\n"
     ]
    },
    {
     "ename": "",
     "evalue": "127",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "\n",
    "#### CD-HIT ####\n",
    "# clustering sequences with cd-hit\n",
    "cdhit_opt=\"-p 1 -c 0.75 -n 4 -T 0 -M 0\"\n",
    "cdhit_nmsfx=\"_75\"\n",
    "\n",
    "# AC\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[0]} -i2 ${path_results}${seq_200bp[1]} -o ${path_results}cd-hit/AC_AS_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[0]} -i2 ${path_results}${seq_200bp[2]} -o ${path_results}cd-hit/AC_EH_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "# AS\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[1]} -i2 ${path_results}${seq_200bp[2]} -o ${path_results}cd-hit/AS_EH_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[1]} -i2 ${path_results}${seq_200bp[0]} -o ${path_results}cd-hit/AS_AC_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "# EH\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[2]} -i2 ${path_results}${seq_200bp[0]} -o ${path_results}cd-hit/EH_AC_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[2]} -i2 ${path_results}${seq_200bp[1]} -o ${path_results}cd-hit/EH_AS_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "\n",
    "\n",
    "# loop test\n",
    ": << 'COMMENT'\n",
    "for i in $seq_200bp; do\n",
    "    cd-hit-est-2d -i ${path_results}$i -i2 ${path_results}AC -o ${path_results}cd-hit/${i}cd-hit_AC.fasta -c 0.9 -p 1\n",
    "done\n",
    "COMMENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bf8e06c-ab45-472d-b55d-0384f38d0bcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### BOWTIE2 ##### \n",
    "#for i in \n",
    "#bowtie2 -x ${path_raw}ZF/rgenome/bowtie2-index/ZF_bowtie2 -f -U ${path_results}EH_CpG_200bp.fasta -S ${path_results}ZF_EH_200_bt2_N_1.sam -N 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a6903d20-9dbf-45ef-8555-3bba82d40e5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "#### MINIMAP2 ####\n",
    "\n",
    "#AC\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_200bp[0]} > ${path_results}minimap2/ZF_AC_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_100bp[0]} > ${path_results}minimap2/ZF_AC_100_minimap.sam\n",
    "#AS\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_200bp[1]} > ${path_results}minimap2/ZF_AS_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_100bp[1]} > ${path_results}minimap2/ZF_AS_100_minimap.sam\n",
    "#EH\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_200bp[2]} > ${path_results}minimap2/ZF_EH_200_minimap.sam \n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_100bp[2]} > ${path_results}minimap2/ZF_EH_100_minimap.sam \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef724657-b0c3-4a76-b119-afe8bcbef7ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "## view how many alignments alignments\n",
    "\n",
    "#AC\n",
    "view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AC_200_minimap.sam\n",
    "view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AC_100_minimap.sam\n",
    "#AS\n",
    "view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AS_200_minimap.sam\n",
    "view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AS_100_minimap.sam\n",
    "#EH\n",
    "view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_EH_200_minimap.sam\n",
    "view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_EH_100_minimap.sam"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
