{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2685f3a6-cf6b-4c7d-bac3-37ca3d5157d2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently Loaded Modulefiles:\n",
      " 1) \u001b[46mpowerPlant/core\u001b[0m    5) \u001b[46mSlurm/21.08.8-2\u001b[0m      9) bowtie2/2.3.4.3  \n",
      " 2) texlive/20230914   6) perlbrew/0.76       10) minimap2/2.22    \n",
      " 3) pandoc/1.19.2      7) perl/5.36.0         11) cd-hit/4.6.1     \n",
      " 4) git/2.21.0         8) slurm-utils/latest  12) samtools/1.16    \n",
      "\n",
      "Key:\n",
      "\u001b[46msticky\u001b[0m  \n"
     ]
    }
   ],
   "source": [
    "# Alignments for various species to reference genomes\n",
    "\n",
    "# loading modules\n",
    "module load bowtie2\n",
    "module load minimap2 \n",
    "module load cd-hit\n",
    "module load samtools\n",
    "\n",
    "# checking active \n",
    "module list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3eec4422-fd71-4360-90d1-1fe9b4aaddfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# setting up variables \n",
    "# paths\n",
    "path_raw=/workspace/cfngle/raw-data/\n",
    "path_results=/workspace/cfngle/results-data/\n",
    "\n",
    "# filenames\n",
    "seq_100bp=(\"AC_CpG_100bp.fasta\" \"AS_CpG_100bp.fasta\" \"EH_CpG_100bp.fasta\")\n",
    "\n",
    "seq_200bp=(\"AC_CpG_200bp.fasta\" \"AS_CpG_200bp.fasta\" \"EH_CpG_200bp.fasta\")\n",
    "\n",
    "seq_500bp=(\"AC_CpG_500bp.fasta\" \"AS_CpG_500bp.fasta\" \"EH_CpG_500bp.fasta\")\n",
    "\n",
    "seq_1000bp=(\"AC_CpG_1000bp.fasta\" \"AS_CpG_1000bp.fasta\" \"EH_CpG_1000bp.fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd4afcc3-a0e0-4fae-a616-e9b8dd291e89",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### CD-HIT ####\n",
    "# clustering sequences with cd-hit\n",
    "cdhit_opt=\"-p 1 -c 0.75 -n 4 -T 0 -M 0\"\n",
    "cdhit_nmsfx=\"_75\"\n",
    "\n",
    "# AC\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[0]} -i2 ${path_results}${seq_200bp[1]} -o ${path_results}cd-hit/AC_AS_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[0]} -i2 ${path_results}${seq_200bp[2]} -o ${path_results}cd-hit/AC_EH_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "# AS\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[1]} -i2 ${path_results}${seq_200bp[2]} -o ${path_results}cd-hit/AS_EH_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[1]} -i2 ${path_results}${seq_200bp[0]} -o ${path_results}cd-hit/AS_AC_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "# EH\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[2]} -i2 ${path_results}${seq_200bp[0]} -o ${path_results}cd-hit/EH_AC_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[2]} -i2 ${path_results}${seq_200bp[1]} -o ${path_results}cd-hit/EH_AS_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "\n",
    "\n",
    "# loop test\n",
    ": << 'COMMENT'\n",
    "for i in $seq_200bp; do\n",
    "    cd-hit-est-2d -i ${path_results}$i -i2 ${path_results}AC -o ${path_results}cd-hit/${i}cd-hit_AC.fasta -c 0.9 -p 1\n",
    "done\n",
    "COMMENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "02f20b45-2815-4c8f-83e8-e57cafa7b3a1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13900 reads; of these:\n",
      "  13900 (100.00%) were unpaired; of these:\n",
      "    12449 (89.56%) aligned 0 times\n",
      "    487 (3.50%) aligned exactly 1 time\n",
      "    964 (6.94%) aligned >1 times\n",
      "10.44% overall alignment rate\n",
      "13755 reads; of these:\n",
      "  13755 (100.00%) were unpaired; of these:\n",
      "    11497 (83.58%) aligned 0 times\n",
      "    542 (3.94%) aligned exactly 1 time\n",
      "    1716 (12.48%) aligned >1 times\n",
      "16.42% overall alignment rate\n",
      "13720 reads; of these:\n",
      "  13720 (100.00%) were unpaired; of these:\n",
      "    10136 (73.88%) aligned 0 times\n",
      "    512 (3.73%) aligned exactly 1 time\n",
      "    3072 (22.39%) aligned >1 times\n",
      "26.12% overall alignment rate\n",
      "13641 reads; of these:\n",
      "  13641 (100.00%) were unpaired; of these:\n",
      "    8586 (62.94%) aligned 0 times\n",
      "    544 (3.99%) aligned exactly 1 time\n",
      "    4511 (33.07%) aligned >1 times\n",
      "37.06% overall alignment rate\n",
      "27560 reads; of these:\n",
      "  27560 (100.00%) were unpaired; of these:\n",
      "    26269 (95.32%) aligned 0 times\n",
      "    643 (2.33%) aligned exactly 1 time\n",
      "    648 (2.35%) aligned >1 times\n",
      "4.68% overall alignment rate\n",
      "24535 reads; of these:\n",
      "  24535 (100.00%) were unpaired; of these:\n",
      "    22832 (93.06%) aligned 0 times\n",
      "    700 (2.85%) aligned exactly 1 time\n",
      "    1003 (4.09%) aligned >1 times\n",
      "6.94% overall alignment rate\n",
      "24489 reads; of these:\n",
      "  24489 (100.00%) were unpaired; of these:\n",
      "    21986 (89.78%) aligned 0 times\n",
      "    794 (3.24%) aligned exactly 1 time\n",
      "    1709 (6.98%) aligned >1 times\n",
      "10.22% overall alignment rate\n",
      "24410 reads; of these:\n",
      "  24410 (100.00%) were unpaired; of these:\n",
      "    21021 (86.12%) aligned 0 times\n",
      "    919 (3.76%) aligned exactly 1 time\n",
      "    2470 (10.12%) aligned >1 times\n",
      "13.88% overall alignment rate\n",
      "20018 reads; of these:\n",
      "  20018 (100.00%) were unpaired; of these:\n",
      "    16588 (82.87%) aligned 0 times\n",
      "    792 (3.96%) aligned exactly 1 time\n",
      "    2638 (13.18%) aligned >1 times\n",
      "17.13% overall alignment rate\n",
      "18337 reads; of these:\n",
      "  18337 (100.00%) were unpaired; of these:\n",
      "    13870 (75.64%) aligned 0 times\n",
      "    782 (4.26%) aligned exactly 1 time\n",
      "    3685 (20.10%) aligned >1 times\n",
      "24.36% overall alignment rate\n",
      "18193 reads; of these:\n",
      "  18193 (100.00%) were unpaired; of these:\n",
      "    12046 (66.21%) aligned 0 times\n",
      "    777 (4.27%) aligned exactly 1 time\n",
      "    5370 (29.52%) aligned >1 times\n",
      "33.79% overall alignment rate\n",
      "18049 reads; of these:\n",
      "  18049 (100.00%) were unpaired; of these:\n",
      "    10311 (57.13%) aligned 0 times\n",
      "    815 (4.52%) aligned exactly 1 time\n",
      "    6923 (38.36%) aligned >1 times\n",
      "42.87% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "## BOWTIE2 ####\n",
    "\n",
    "## modules\n",
    "module load bowtie\n",
    "\n",
    "## setting up variables \n",
    "# paths\n",
    "path_raw=/workspace/cfngle/raw-data/\n",
    "path_results=/workspace/cfngle/results-data/bowtie2/\n",
    "path_sequences=/workspace/cfngle/results-data/sequences/\n",
    "\n",
    "# filenames\n",
    "seq_100bp=(\"AC_CpG_100bp\" \"AS_CpG_100bp\" \"EH_CpG_100bp\")\n",
    "seq_200bp=(\"AC_CpG_200bp\" \"AS_CpG_200bp\" \"EH_CpG_200bp\")\n",
    "seq_500bp=(\"AC_CpG_500bp\" \"AS_CpG_500bp\" \"EH_CpG_500bp\")\n",
    "seq_1000bp=(\"AC_CpG_1000bp\" \"AS_CpG_1000bp\" \"EH_CpG_1000bp\")\n",
    "\n",
    "# file ending\n",
    "suffix=\".fasta\"\n",
    "\n",
    "# arguments\n",
    "bowtie2_args=\"--very-sensitive --local -p 8\"\n",
    "\n",
    "\n",
    "## rgenome ZF\n",
    "echo \"ZF\"\n",
    "for (( i=0; i<${#seq_100bp[@]}; i++ )); \n",
    "do \n",
    "   bowtie2 $bowtie2_args -x ${path_raw}ZF/rgenome/bowtie2-index/ZF_bowtie2 -f -U ${path_sequences}${seq_100bp[$i]}$suffix -S ${path_results}ZF_${seq_100bp[$i]}_bt2_.sam -N 1\n",
    "   bowtie2 $bowtie2_args -x ${path_raw}ZF/rgenome/bowtie2-index/ZF_bowtie2 -f -U ${path_sequences}${seq_200bp[$i]}$suffix -S ${path_results}ZF_${seq_200bp[$i]}_bt2_.sam -N 1\n",
    "   bowtie2 $bowtie2_args -x ${path_raw}ZF/rgenome/bowtie2-index/ZF_bowtie2 -f -U ${path_sequences}${seq_500bp[$i]}$suffix -S ${path_results}ZF_${seq_500bp[$i]}_bt2_.sam -N 1\n",
    "   bowtie2 $bowtie2_args -x ${path_raw}ZF/rgenome/bowtie2-index/ZF_bowtie2 -f -U ${path_sequences}${seq_1000bp[$i]}$suffix -S ${path_results}ZF_${seq_1000bp[$i]}_bt2_.sam -N 1\n",
    "done\n",
    "\n",
    "## rgenome AC\n",
    "echo \"AC\"\n",
    "for (( i=0; i<${#seq_100bp[@]}; i++ )); \n",
    "do \n",
    "    bowtie2 $bowtie2_args -x ${path_raw}AC/bowtie2-index/AC_bowtie2 -f -U ${path_sequences}${seq_100bp[$i]}$suffix -S ${path_results}AC_${seq_100bp[$i]}_bt2_.sam -N 1\n",
    "    bowtie2 $bowtie2_args -x ${path_raw}AC/bowtie2-index/AC_bowtie2 -f -U ${path_sequences}${seq_200bp[$i]}$suffix -S ${path_results}AC_${seq_200bp[$i]}_bt2_.sam -N 1\n",
    "    bowtie2 $bowtie2_args -x ${path_raw}AC/bowtie2-index/AC_bowtie2 -f -U ${path_sequences}${seq_500bp[$i]}$suffix -S ${path_results}AC_${seq_500bp[$i]}_bt2_.sam -N 1\n",
    "    bowtie2 $bowtie2_args -x ${path_raw}AC/bowtie2-index/AC_bowtie2 -f -U ${path_sequences}${seq_1000bp[$i]}$suffix -S ${path_results}AC_${seq_1000bp[$i]}_bt2_.sam -N 1\n",
    "done\n",
    "\n",
    "## rgenome AS\n",
    "echo \"AS\"\n",
    "for (( i=0; i<${#seq_100bp[@]}; i++ )); \n",
    "do \n",
    "    bowtie2 $bowtie2_args -x ${path_raw}AS/bowtie2-index/AS_bowtie2 -f -U ${path_sequences}${seq_100bp[$i]}$suffix -S ${path_results}AS_${seq_100bp[$i]}_bt2_.sam -N 1\n",
    "    bowtie2 $bowtie2_args -x ${path_raw}AS/bowtie2-index/AS_bowtie2 -f -U ${path_sequences}${seq_200bp[$i]}$suffix -S ${path_results}AS_${seq_200bp[$i]}_bt2_.sam -N 1\n",
    "    bowtie2 $bowtie2_args -x ${path_raw}AS/bowtie2-index/AS_bowtie2 -f -U ${path_sequences}${seq_500bp[$i]}$suffix -S ${path_results}AS_${seq_500bp[$i]}_bt2_.sam -N 1\n",
    "    bowtie2 $bowtie2_args -x ${path_raw}AS/bowtie2-index/AS_bowtie2 -f -U ${path_sequences}${seq_1000bp[$i]}$suffix -S ${path_results}AS_${seq_1000bp[$i]}_bt2_.sam -N 1\n",
    "done\n",
    "\n",
    "## rgenome EH\n",
    "echo \"EH\"\n",
    "for (( i=0; i<${#seq_100bp[@]}; i++ )); \n",
    "do \n",
    "    bowtie2 $bowtie2_args -x ${path_raw}EH/bowtie2-index/EH_bowtie2 -f -U ${path_sequences}${seq_100bp[$i]}$suffix -S ${path_results}EH_${seq_100bp[$i]}_bt2_.sam -N 1\n",
    "    bowtie2 $bowtie2_args -x ${path_raw}EH/bowtie2-index/EH_bowtie2 -f -U ${path_sequences}${seq_200bp[$i]}$suffix -S ${path_results}EH_${seq_200bp[$i]}_bt2_.sam -N 1\n",
    "    bowtie2 $bowtie2_args -x ${path_raw}EH/bowtie2-index/EH_bowtie2 -f -U ${path_sequences}${seq_500bp[$i]}$suffix -S ${path_results}EH_${seq_500bp[$i]}_bt2_.sam -N 1\n",
    "    bowtie2 $bowtie2_args -x ${path_raw}EH/bowtie2-index/EH_bowtie2 -f -U ${path_sequences}${seq_1000bp[$i]}$suffix -S ${path_results}EH_${seq_1000bp[$i]}_bt2_.sam -N 1\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "2183866e-6182-4acc-985b-0a4a03276aa4",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"/workspace/cfngle/raw-data/AS/bowtie2-index/AS_bowtie2.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /workspace/cfngle/raw-data/AS/Chrysophrys_auratus.v.1.0.all.male.map.fasta\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:06\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:04\n",
      "bmax according to bmaxDivN setting: 176668623\n",
      "Using parameters --bmax 132501468 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 132501468 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:15\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:04\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:07\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 7.06674e+08 (target: 132501467)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 706674494 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:07:18\n",
      "Returning block of 706674495 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 208442737\n",
      "fchr[G]: 353300318\n",
      "fchr[T]: 498179311\n",
      "fchr[$]: 706674494\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 240273095 bytes to primary EBWT file: /workspace/cfngle/raw-data/AS/bowtie2-index/AS_bowtie2.1.bt2\n",
      "Wrote 176668628 bytes to secondary EBWT file: /workspace/cfngle/raw-data/AS/bowtie2-index/AS_bowtie2.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 706674494\n",
      "    bwtLen: 706674495\n",
      "    sz: 176668624\n",
      "    bwtSz: 176668624\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 44167156\n",
      "    offsSz: 176668624\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 3680597\n",
      "    numLines: 3680597\n",
      "    ebwtTotLen: 235558208\n",
      "    ebwtTotSz: 235558208\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:09:46\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:03\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:04\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 176668623\n",
      "Using parameters --bmax 132501468 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 132501468 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:15\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:04\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:07\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 7.06674e+08 (target: 132501467)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 706674494 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:07:05\n",
      "Returning block of 706674495 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 208442737\n",
      "fchr[G]: 353300318\n",
      "fchr[T]: 498179311\n",
      "fchr[$]: 706674494\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 240273095 bytes to primary EBWT file: /workspace/cfngle/raw-data/AS/bowtie2-index/AS_bowtie2.rev.1.bt2\n",
      "Wrote 176668628 bytes to secondary EBWT file: /workspace/cfngle/raw-data/AS/bowtie2-index/AS_bowtie2.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 706674494\n",
      "    bwtLen: 706674495\n",
      "    sz: 176668624\n",
      "    bwtSz: 176668624\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 44167156\n",
      "    offsSz: 176668624\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 3680597\n",
      "    numLines: 3680597\n",
      "    ebwtTotLen: 235558208\n",
      "    ebwtTotSz: 235558208\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:09:30\n"
     ]
    }
   ],
   "source": [
    "#### Bowtie2 indexing ####\n",
    "## modules\n",
    "module load bowtie\n",
    "\n",
    "## setting up variables \n",
    "# paths\n",
    "path_raw=/workspace/cfngle/raw-data/\n",
    "path_results=/workspace/cfngle/results-data/bowtie2/\n",
    "path_sequences=/workspace/cfngle/results-data/sequences/\n",
    "\n",
    "# filenames\n",
    "seq_100bp=(\"AC_CpG_100bp\" \"AS_CpG_100bp\" \"EH_CpG_100bp\")\n",
    "seq_200bp=(\"AC_CpG_200bp\" \"AS_CpG_200bp\" \"EH_CpG_200bp\")\n",
    "seq_500bp=(\"AC_CpG_500bp\" \"AS_CpG_500bp\" \"EH_CpG_500bp\")\n",
    "seq_1000bp=(\"AC_CpG_1000bp\" \"AS_CpG_1000bp\" \"EH_CpG_1000bp\")\n",
    "\n",
    "## ZF\n",
    "#bowtie2-build ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_raw}ZF/rgenome/bowtie2-index/ZF_bowtie2\n",
    "## AC\n",
    "#bowtie2-build ${path_raw}AC/GCF_902167405.1_gadMor3.0_genomic.fasta ${path_raw}AC/bowtie2-index/AC_bowtie2\n",
    "## AS\n",
    "#bowtie2-build ${path_raw}AS/Chrysophrys_auratus.v.1.0.all.male.map.fasta ${path_raw}AS/bowtie2-index/AS_bowtie2\n",
    "## EH\n",
    "#bowtie2-build ${path_raw}EH/fMerMel2.1_cnag1.scaffolds.fa ${path_raw}EH/bowtie2-index/EH_bowtie2\n",
    "## JM\n",
    "#bowtie2-build ${path_raw}JM/rgenome/GCF_002234675.1_ASM223467v1_genomic.fna ${path_raw}JM/rgenome/bowtie2-index/JM_bowtie2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a6903d20-9dbf-45ef-8555-3bba82d40e5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "#### MINIMAP2 ####\n",
    "\n",
    "#ZF\n",
    "#AC\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_200bp[0]} > ${path_results}minimap2/ZF_AC_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_100bp[0]} > ${path_results}minimap2/ZF_AC_100_minimap.sam\n",
    "#AS\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_200bp[1]} > ${path_results}minimap2/ZF_AS_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_100bp[1]} > ${path_results}minimap2/ZF_AS_100_minimap.sam\n",
    "#EH\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_200bp[2]} > ${path_results}minimap2/ZF_EH_200_minimap.sam \n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_100bp[2]} > ${path_results}minimap2/ZF_EH_100_minimap.sam \n",
    "\n",
    "path_rgenome=\"AC/GCF_902167405.1_gadMor3.0_genomic.fasta\"\n",
    "#AC\n",
    "#AC\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[0]} > ${path_results}minimap2/AC_AC_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[0]} > ${path_results}minimap2/AC_AC_100_minimap.sam\n",
    "#AS\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[1]} > ${path_results}minimap2/AC_AS_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[1]} > ${path_results}minimap2/AC_AS_100_minimap.sam\n",
    "#EH\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[2]} > ${path_results}minimap2/AC_EH_200_minimap.sam \n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[2]} > ${path_results}minimap2/AC_EH_100_minimap.sam \n",
    "\n",
    "path_rgenome=\"Chrysophrys_auratus.v.1.0.all.male.map.fasta\"\n",
    "#AS\n",
    "#AC\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[0]} > ${path_results}minimap2/AS_AC_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[0]} > ${path_results}minimap2/AS_AC_100_minimap.sam\n",
    "#AS\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[1]} > ${path_results}minimap2/AS_AS_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[1]} > ${path_results}minimap2/AS_AS_100_minimap.sam\n",
    "#EH\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[2]} > ${path_results}minimap2/AS_EH_200_minimap.sam \n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[2]} > ${path_results}minimap2/AS_EH_100_minimap.sam \n",
    "\n",
    "path_rgenome=\"fMerMel2.1_cnag1.scaffolds.fa\"\n",
    "#EH\\\n",
    "#AC\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[0]} > ${path_results}minimap2/EH_AC_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[0]} > ${path_results}minimap2/EH_AC_100_minimap.sam\n",
    "#AS\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[1]} > ${path_results}minimap2/EH_AS_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[1]} > ${path_results}minimap2/EH_AS_100_minimap.sam\n",
    "#EH\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[2]} > ${path_results}minimap2/EH_EH_200_minimap.sam \n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[2]} > ${path_results}minimap2/EH_EH_100_minimap.sam \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ef724657-b0c3-4a76-b119-afe8bcbef7ad",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "173\n",
      "415\n",
      "1391\n",
      "2680\n",
      "265\n",
      "497\n",
      "1081\n",
      "1975\n",
      "456\n",
      "926\n",
      "2863\n",
      "5472\n",
      "16349\n",
      "15460\n",
      "14918\n",
      "14480\n",
      "1397\n",
      "2073\n",
      "3218\n",
      "4906\n",
      "10165\n",
      "12835\n",
      "18254\n",
      "25553\n",
      "844\n",
      "1231\n",
      "2149\n",
      "3650\n",
      "28688\n",
      "24797\n",
      "24642\n",
      "24530\n",
      "2261\n",
      "3698\n",
      "5642\n",
      "8460\n",
      "5126\n",
      "7118\n",
      "12052\n",
      "18420\n",
      "1588\n",
      "2508\n",
      "4191\n",
      "6812\n",
      "24022\n",
      "21442\n",
      "20341\n",
      "19491\n"
     ]
    }
   ],
   "source": [
    "## view how many alignments (multiple alignments)\n",
    "\n",
    "##ZF\n",
    "#AC\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AC_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AC_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AC_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AC_1000_minimap.sam\n",
    "#AS\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AS_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AS_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AS_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AS_1000_minimap.sam\n",
    "#EH\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_EH_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_EH_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_EH_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_EH_1000_minimap.sam\n",
    "\n",
    "##AC\n",
    "#AC\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AC_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AC_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AC_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AC_1000_minimap.sam\n",
    "#AS\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AS_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AS_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AS_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AS_1000_minimap.sam\n",
    "#EH\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_EH_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_EH_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_EH_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_EH_1000_minimap.sam\n",
    "\n",
    "##AS\n",
    "#AC\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AC_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AC_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AC_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AC_1000_minimap.sam\n",
    "#AS\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AS_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AS_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AS_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AS_1000_minimap.sam\n",
    "#EH\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_EH_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_EH_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_EH_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_EH_1000_minimap.sam\n",
    "\n",
    "##EH\n",
    "#AC\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AC_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AC_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AC_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AC_1000_minimap.sam\n",
    "#AS\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AS_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AS_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AS_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AS_1000_minimap.sam\n",
    "#EH\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_EH_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_EH_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_EH_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_EH_1000_minimap.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "05689c53-68b5-42de-bd55-bf1a229e93b6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "107\n",
      "207\n",
      "515\n",
      "887\n",
      "158\n",
      "280\n",
      "543\n",
      "896\n",
      "268\n",
      "430\n",
      "970\n",
      "1662\n",
      "13879\n",
      "13755\n",
      "13720\n",
      "13641\n",
      "802\n",
      "1165\n",
      "1861\n",
      "2809\n",
      "6253\n",
      "7767\n",
      "10012\n",
      "12323\n",
      "514\n",
      "768\n",
      "1277\n",
      "2051\n",
      "27538\n",
      "24535\n",
      "24489\n",
      "24410\n",
      "1252\n",
      "1768\n",
      "2666\n",
      "3912\n",
      "3888\n",
      "5192\n",
      "7301\n",
      "9543\n",
      "958\n",
      "1467\n",
      "2493\n",
      "3817\n",
      "19904\n",
      "18329\n",
      "18192\n",
      "18049\n"
     ]
    }
   ],
   "source": [
    "# view single alignments\n",
    "\n",
    "##ZF\n",
    "#AC\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AC_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AC_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AC_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AC_1000_minimap.sam | wc -l\n",
    "#AS\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AS_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AS_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AS_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AS_1000_minimap.sam | wc -l\n",
    "#EH\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_EH_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_EH_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_EH_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_EH_1000_minimap.sam | wc -l\n",
    "\n",
    "##AC\n",
    "#AC\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AC_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AC_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AC_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AC_1000_minimap.sam | wc -l\n",
    "#AS\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AS_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AS_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AS_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AS_1000_minimap.sam | wc -l\n",
    "#EH\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_EH_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_EH_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_EH_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_EH_1000_minimap.sam | wc -l\n",
    "\n",
    "##AS\n",
    "#AC\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AC_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AC_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AC_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AC_1000_minimap.sam | wc -l\n",
    "#AS\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AS_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AS_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AS_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AS_1000_minimap.sam | wc -l\n",
    "#EH\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_EH_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_EH_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_EH_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_EH_1000_minimap.sam | wc -l\n",
    "\n",
    "##EH\n",
    "#AC\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AC_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AC_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AC_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AC_1000_minimap.sam | wc -l\n",
    "#AS\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AS_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AS_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AS_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AS_1000_minimap.sam | wc -l\n",
    "#EH\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_EH_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_EH_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_EH_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_EH_1000_minimap.sam | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ce76fb89-78a8-44db-811a-be886dd467cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "single alignment hits\n",
      "1373\n",
      "1461\n",
      "4868\n",
      "2152\n",
      "26351\n",
      "multi alignment hits\n",
      "5167\n",
      "3109\n",
      "8031\n",
      "4935\n",
      "27890\n"
     ]
    }
   ],
   "source": [
    "module load samtools\n",
    "echo \"single alignment hits\"\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_JM_1000_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_JM_1000_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_JM_1000_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_JM_1000_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/JM_JM_1000_minimap.sam | wc -l\n",
    "echo \"multi alignment hits\"\n",
    "samtools view -F 4 /workspace/cfngle/results-data/minimap2/ZF_JM_1000_minimap.sam -c\n",
    "samtools view -F 4 /workspace/cfngle/results-data/minimap2/AC_JM_1000_minimap.sam -c\n",
    "samtools view -F 4 /workspace/cfngle/results-data/minimap2/AS_JM_1000_minimap.sam -c\n",
    "samtools view -F 4 /workspace/cfngle/results-data/minimap2/EH_JM_1000_minimap.sam -c\n",
    "samtools view -F 4 /workspace/cfngle/results-data/minimap2/JM_JM_1000_minimap.sam -c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3eafa8c1-2bc0-4d30-ad71-4e09ca7bbb37",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::32.235*1.63] collected minimizers\n",
      "[M::mm_idx_gen::40.264*1.89] sorted minimizers\n",
      "[M::main::40.266*1.89] loaded/built the index for 1923 target sequence(s)\n",
      "[M::mm_mapopt_update::41.644*1.86] mid_occ = 848\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1923\n",
      "[M::mm_idx_stat::42.496*1.84] distinct minimizers: 66057654 (43.51% are singletons); average occurrences: 4.867; average spacing: 5.223; total length: 1679203469\n",
      "[M::worker_pipeline::66.027*2.21] mapped 16751 sequences\n",
      "[M::main] Version: 2.22-r1101\n",
      "[M::main] CMD: minimap2 -ax map-ont /workspace/cfngle/raw-data/ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna /workspace/cfngle/results-data/sequences/EH_CpG_5000bp.fasta\n",
      "[M::main] Real time: 66.285 sec; CPU: 146.265 sec; Peak RSS: 6.608 GB\n"
     ]
    }
   ],
   "source": [
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_sequences}EH_CpG_5000bp.fasta > ${path_results}minimap2/ZF_EH_5000_minimap.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6a2dff7d-0080-4bfd-909b-726e2f3eefd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5557\n"
     ]
    }
   ],
   "source": [
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_EH_5000_minimap.sam | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "f18465a1-e736-45c5-8a1a-9f8c8355800b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OS = linux\n",
      "The number of physical cores =  10\n",
      "nthread = 10\n",
      "stacksize: 8192 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 1751 ambiguous characters.\n",
      "  401 / 402 (thread    6)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "  400 / 402\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP   124 / 401 (thread    4)df\n",
      "Reallocating..done. *alloclen = 6014\n",
      "STEP   252 / 401 (thread    6)d\n",
      "Reallocating..done. *alloclen = 7251\n",
      "STEP   351 / 401 (thread    7)d\n",
      "Reallocating..done. *alloclen = 10612\n",
      "STEP   360 / 401 (thread    9)d\n",
      "Reallocating..done. *alloclen = 12970\n",
      "STEP   364 / 401 (thread    0)f\n",
      "Reallocating..done. *alloclen = 15349\n",
      "STEP   363 / 401 (thread    9)d\n",
      "Reallocating..done. *alloclen = 16824\n",
      "STEP   401 / 401 (thread    7)d\n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "  400 / 402 (thread    4)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "  400 / 402\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP   160 / 401 (thread    3)f\n",
      "Reallocating..done. *alloclen = 5791\n",
      "STEP   292 / 401 (thread    2)d\n",
      "Reallocating..done. *alloclen = 7741\n",
      "STEP   356 / 401 (thread    6)d\n",
      "Reallocating..done. *alloclen = 10061\n",
      "STEP   373 / 401 (thread    2)d\n",
      "Reallocating..done. *alloclen = 15183\n",
      "STEP   378 / 401 (thread    8)d\n",
      "Reallocating..done. *alloclen = 16936\n",
      "STEP   395 / 401 (thread    9)f\n",
      "Reallocating..done. *alloclen = 20028\n",
      "STEP   401 / 401 (thread    1)d\n",
      "done.\n",
      "\n",
      "disttbfast (nuc) Version 7.307 alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "10 thread(s)\n",
      "\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "  400 / 402 (thread    8)dndpre (nuc) Version 7.307 alg=X, model=DNA200 (2), 1.53 (4.59), 0.37 (1.11), noshift, amax=0.0\n",
      "10 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 10\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "\n",
      "  400 / 402\n",
      "Segment   1/  1    1-15103\n",
      "002-0653-0 (thread   10) better        \n"
     ]
    }
   ],
   "source": [
    "#### MAFFT ####\n",
    "\n",
    "## modules\n",
    "module load mafft\n",
    "\n",
    "## setting up variables \n",
    "\n",
    "# paths\n",
    "path_raw=/workspace/cfngle/raw-data/\n",
    "path_results=/workspace/cfngle/results-data/\n",
    "path_sequences=/workspace/cfngle/results-data/sequences/\n",
    "\n",
    "# arguments\n",
    "mafft_args=\"--auto --thread -1\"\n",
    "\n",
    "mafft $mafft_args ${path_results}02_conserved_seq/AC_AS_EH_1000_conserved.fasta > ${path_results}02_conserved_seq/AC_AS_EH_1000_conserved_MSA_mafft.fasta\n",
    "#mafft $mafft_args ${path_sequences}AC_AS_EH_CpG_100bp.fasta > ${path_results}AC_AS_EH_CpG_100bp.fasta \n",
    "#mafft $mafft_args ${path_sequences}AC_AS_EH_CpG_200bp.fasta > ${path_results}AC_AS_EH_CpG_200bp.fasta \n",
    "#mafft $mafft_args ${path_sequences}AC_AS_EH_CpG_500bp.fasta > ${path_results}AC_AS_EH_CpG_500bp.fasta \n",
    "#mafft $mafft_args ${path_sequences}AC_AS_EH_CpG_1000bp.fasta > ${path_results}AC_AS_EH_CpG_1000bp.fasta "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dbe01c7-dd09-4dd5-abdc-2cd289395391",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
