{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4d1f29d1-809b-4e6b-8aac-79fc09282f80",
   "metadata": {},
   "source": [
    "# 00a RRBS data analysis JM\n",
    "\n",
    "\n",
    "Author: Gabriel Ecker Eckhofen (adapted from Dafni Anastasiadi)\n",
    "Date: Febuary 2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f57139d3-47f2-44f6-8491-12363f3603d3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "module load conda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "58a14287-df70-48e9-84ce-23498633f59b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Setting up conda envirnment and activating it\n",
    "conda deactivate\n",
    "conda activate cfngle_env-01\n",
    "\n",
    "# must be done in terminal \n",
    "# conda install -c bioconda entrez-direct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "af2dd13e-3345-467c-8874-ea7f9dcb08dd",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) SAMN21040263\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040262\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040261\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specifiedtput: \n",
      "No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040260\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040259\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040258\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040257\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: tput: No value for $TERM and no -T specifiedNo value for $TERM and no -T specified\n",
      "\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040256\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040255\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040254\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040253\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040252\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040251\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040250\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040249\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040248\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040247\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040246\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040245\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040244\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040243\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040242\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040241\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN21040240\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472656\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472655\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472654\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472653\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472652\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472651\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472650\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472649\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472648\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472647\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472646\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472645\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472644\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472643\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472642\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472641\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472640\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472639\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472638\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472637\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472636\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472635\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "SAMN18472634\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "(/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "cd /workspace/cfngle/raw-data/JM/raw-reads/00_metadata\n",
    "\n",
    "# Fetch metadata for a specific BioSample ID\n",
    "filenames=(\"SAMN21040240.fastq\" \"SAMN21040241.fastq\" \"SAMN21040242.fastq\" \"SAMN21040243.fastq\" \"SAMN21040244.fastq\" \"SAMN21040245.fastq\" \"SAMN21040246.fastq\" \n",
    "\"SAMN21040247.fastq\" \"SAMN21040248.fastq\" \"SAMN21040249.fastq\" \"SAMN21040250.fastq\" \"SAMN21040251.fastq\" \"SAMN21040252.fastq\" \"SAMN21040253.fastq\" \"SAMN21040254.fastq\" \n",
    "\"SAMN21040255.fastq\" \"SAMN21040256.fastq\" \"SAMN21040257.fastq\" \"SAMN21040258.fastq\" \"SAMN21040259.fastq\" \"SAMN21040260.fastq\" \"SAMN21040261.fastq\" \"SAMN21040262.fastq\" \"SAMN21040263.fastq\" \n",
    "\"SRR18462633.fastq\" \"SRR18462634.fastq\" \"SRR18462635.fastq\" \"SRR18462636.fastq\" \"SRR18462637.fastq\" \"SRR18462638.fastq\" \"SRR18462639.fastq\" \"SRR18462640.fastq\" \n",
    "\"SRR18462641.fastq\" \"SRR18462642.fastq\" \"SRR18462643.fastq\" \"SRR18462644.fastq\" \"SRR18462645.fastq\" \"SRR18462646.fastq\" \"SRR18462647.fastq\" \"SRR18462648.fastq\" \n",
    "\"SRR18462649.fastq\" \"SRR18462650.fastq\" \"SRR18462651.fastq\" \"SRR18462652.fastq\" \"SRR18462653.fastq\" \"SRR18462654.fastq\" \"SRR18462655.fastq\" )\n",
    "\n",
    "filenames=(\"SAMN21040263\" \"SAMN21040262\" \"SAMN21040261\" \"SAMN21040260\" \"SAMN21040259\" \"SAMN21040258\" \"SAMN21040257\" \"SAMN21040256\"\n",
    "\"SAMN21040255\" \"SAMN21040254\" \"SAMN21040253\" \"SAMN21040252\" \"SAMN21040251\" \"SAMN21040250\" \"SAMN21040249\" \"SAMN21040248\"\n",
    "\"SAMN21040247\" \"SAMN21040246\" \"SAMN21040245\" \"SAMN21040244\" \"SAMN21040243\" \"SAMN21040242\" \"SAMN21040241\" \"SAMN21040240\"\n",
    "\"SAMN18472656\" \"SAMN18472655\" \"SAMN18472654\" \"SAMN18472653\" \"SAMN18472652\" \"SAMN18472651\" \"SAMN18472650\" \"SAMN18472649\"\n",
    "\"SAMN18472648\" \"SAMN18472647\" \"SAMN18472646\" \"SAMN18472645\" \"SAMN18472644\" \"SAMN18472643\" \"SAMN18472642\" \"SAMN18472641\"\n",
    "\"SAMN18472640\" \"SAMN18472639\" \"SAMN18472638\" \"SAMN18472637\" \"SAMN18472636\" \"SAMN18472635\" \"SAMN18472634\")\n",
    "\n",
    "for (( i=0; i<${#filenames[@]}; i++ )); do\n",
    "    # Index\n",
    "    echo \"${filenames[i]%.fastq}\"\n",
    "    esearch -db biosample -query ${filenames[i]%.fastq} | efetch -format native > ${filenames[i]}.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bde928db-5685-43df-b3b6-4b53d96bdd4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda deactivate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c1a0dbeb-7b6f-4cad-890e-01bd6ee3f2a1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) SAMN21040240.fastq\n",
      "JM_001.fastq\n",
      "SAMN21040241.fastq\n",
      "JM_002.fastq\n",
      "SAMN21040242.fastq\n",
      "JM_003.fastq\n",
      "SAMN21040243.fastq\n",
      "JM_004.fastq\n",
      "SAMN21040244.fastq\n",
      "JM_005.fastq\n",
      "SAMN21040245.fastq\n",
      "JM_006.fastq\n",
      "SAMN21040246.fastq\n",
      "JM_007.fastq\n",
      "SAMN21040247.fastq\n",
      "JM_008.fastq\n",
      "SAMN21040248.fastq\n",
      "JM_009.fastq\n",
      "SAMN21040249.fastq\n",
      "JM_010.fastq\n",
      "SAMN21040250.fastq\n",
      "JM_011.fastq\n",
      "SAMN21040251.fastq\n",
      "JM_012.fastq\n",
      "SAMN21040252.fastq\n",
      "JM_013.fastq\n",
      "SAMN21040253.fastq\n",
      "JM_014.fastq\n",
      "SAMN21040254.fastq\n",
      "JM_015.fastq\n",
      "SAMN21040255.fastq\n",
      "JM_016.fastq\n",
      "SAMN21040256.fastq\n",
      "JM_017.fastq\n",
      "SAMN21040257.fastq\n",
      "JM_018.fastq\n",
      "SAMN21040258.fastq\n",
      "JM_019.fastq\n",
      "SAMN21040259.fastq\n",
      "JM_020.fastq\n",
      "SAMN21040260.fastq\n",
      "JM_021.fastq\n",
      "SAMN21040261.fastq\n",
      "JM_022.fastq\n",
      "SAMN21040262.fastq\n",
      "JM_023.fastq\n",
      "SAMN21040263.fastq\n",
      "JM_024.fastq\n",
      "SRR18462633.fastq\n",
      "JM_025.fastq\n",
      "SRR18462634.fastq\n",
      "JM_026.fastq\n",
      "SRR18462635.fastq\n",
      "JM_027.fastq\n",
      "SRR18462636.fastq\n",
      "JM_028.fastq\n",
      "SRR18462637.fastq\n",
      "JM_029.fastq\n",
      "SRR18462638.fastq\n",
      "JM_030.fastq\n",
      "SRR18462639.fastq\n",
      "JM_031.fastq\n",
      "SRR18462640.fastq\n",
      "JM_032.fastq\n",
      "SRR18462641.fastq\n",
      "JM_033.fastq\n",
      "SRR18462642.fastq\n",
      "JM_034.fastq\n",
      "SRR18462643.fastq\n",
      "JM_035.fastq\n",
      "SRR18462644.fastq\n",
      "JM_036.fastq\n",
      "SRR18462645.fastq\n",
      "JM_037.fastq\n",
      "SRR18462646.fastq\n",
      "JM_038.fastq\n",
      "SRR18462647.fastq\n",
      "JM_039.fastq\n",
      "SRR18462648.fastq\n",
      "JM_040.fastq\n",
      "SRR18462649.fastq\n",
      "JM_041.fastq\n",
      "SRR18462650.fastq\n",
      "JM_042.fastq\n",
      "SRR18462651.fastq\n",
      "JM_043.fastq\n",
      "SRR18462652.fastq\n",
      "JM_044.fastq\n",
      "SRR18462653.fastq\n",
      "JM_045.fastq\n",
      "SRR18462654.fastq\n",
      "JM_046.fastq\n",
      "SRR18462655.fastq\n",
      "JM_047.fastq\n",
      "(/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# renaming \n",
    "counter=1\n",
    "for file in *.fastq; do\n",
    "    newname=$(printf \"JM_%03d.fastq\" \"$counter\")\n",
    "    mv \"$file\" \"$newname\"\n",
    "    echo \"$file\"\n",
    "    echo \"$newname\"\n",
    "    ((counter++))\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e3b4904d-4e6c-4f68-84e5-66a992686127",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) filelist.txt\n",
      "JM_001_metadata.txt\n",
      "SAMN21040240.fastq.txt\n",
      "JM_002_metadata.txt\n",
      "SAMN21040241.fastq.txt\n",
      "JM_003_metadata.txt\n",
      "SAMN21040242.fastq.txt\n",
      "JM_004_metadata.txt\n",
      "SAMN21040243.fastq.txt\n",
      "JM_005_metadata.txt\n",
      "SAMN21040244.fastq.txt\n",
      "JM_006_metadata.txt\n",
      "SAMN21040245.fastq.txt\n",
      "JM_007_metadata.txt\n",
      "SAMN21040246.fastq.txt\n",
      "JM_008_metadata.txt\n",
      "SAMN21040247.fastq.txt\n",
      "JM_009_metadata.txt\n",
      "SAMN21040248.fastq.txt\n",
      "JM_010_metadata.txt\n",
      "SAMN21040249.fastq.txt\n",
      "JM_011_metadata.txt\n",
      "SAMN21040250.fastq.txt\n",
      "JM_012_metadata.txt\n",
      "SAMN21040251.fastq.txt\n",
      "JM_013_metadata.txt\n",
      "SAMN21040252.fastq.txt\n",
      "JM_014_metadata.txt\n",
      "SAMN21040253.fastq.txt\n",
      "JM_015_metadata.txt\n",
      "SAMN21040254.fastq.txt\n",
      "JM_016_metadata.txt\n",
      "SAMN21040255.fastq.txt\n",
      "JM_017_metadata.txt\n",
      "SAMN21040256.fastq.txt\n",
      "JM_018_metadata.txt\n",
      "SAMN21040257.fastq.txt\n",
      "JM_019_metadata.txt\n",
      "SAMN21040258.fastq.txt\n",
      "JM_020_metadata.txt\n",
      "SAMN21040259.fastq.txt\n",
      "JM_021_metadata.txt\n",
      "SAMN21040260.fastq.txt\n",
      "JM_022_metadata.txt\n",
      "SAMN21040261.fastq.txt\n",
      "JM_023_metadata.txt\n",
      "SAMN21040262.fastq.txt\n",
      "JM_024_metadata.txt\n",
      "SAMN21040263.fastq.txt\n",
      "JM_025_metadata.txt\n",
      "SRR18462633.fastq.txt\n",
      "JM_026_metadata.txt\n",
      "SRR18462634.fastq.txt\n",
      "JM_027_metadata.txt\n",
      "SRR18462635.fastq.txt\n",
      "JM_028_metadata.txt\n",
      "SRR18462636.fastq.txt\n",
      "JM_029_metadata.txt\n",
      "SRR18462637.fastq.txt\n",
      "JM_030_metadata.txt\n",
      "SRR18462638.fastq.txt\n",
      "JM_031_metadata.txt\n",
      "SRR18462639.fastq.txt\n",
      "JM_032_metadata.txt\n",
      "SRR18462640.fastq.txt\n",
      "JM_033_metadata.txt\n",
      "SRR18462641.fastq.txt\n",
      "JM_034_metadata.txt\n",
      "SRR18462642.fastq.txt\n",
      "JM_035_metadata.txt\n",
      "SRR18462643.fastq.txt\n",
      "JM_036_metadata.txt\n",
      "SRR18462644.fastq.txt\n",
      "JM_037_metadata.txt\n",
      "SRR18462645.fastq.txt\n",
      "JM_038_metadata.txt\n",
      "SRR18462646.fastq.txt\n",
      "JM_039_metadata.txt\n",
      "SRR18462647.fastq.txt\n",
      "JM_040_metadata.txt\n",
      "SRR18462648.fastq.txt\n",
      "JM_041_metadata.txt\n",
      "SRR18462649.fastq.txt\n",
      "JM_042_metadata.txt\n",
      "SRR18462650.fastq.txt\n",
      "JM_043_metadata.txt\n",
      "SRR18462651.fastq.txt\n",
      "JM_044_metadata.txt\n",
      "SRR18462652.fastq.txt\n",
      "JM_045_metadata.txt\n",
      "SRR18462653.fastq.txt\n",
      "JM_046_metadata.txt\n",
      "SRR18462654.fastq.txt\n",
      "JM_047_metadata.txt\n",
      "SRR18462655.fastq.txt\n",
      "JM_048_metadata.txt\n",
      "(/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# renaming \n",
    "counter=1\n",
    "for file in *.txt; do\n",
    "    newname=$(printf \"JM_%03d_metadata.txt\" \"$counter\")\n",
    "    mv \"$file\" \"$newname\"\n",
    "    echo \"$file\"\n",
    "    echo \"$newname\"\n",
    "    ((counter++))\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "756caf22-eda3-4bd3-baf1-e4a0d3038865",
   "metadata": {},
   "source": [
    "## 1. Prepare data and directories\n",
    "\n",
    "Data was downloaded with SRAtoolkit "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "561d77eb-30d4-41e5-8254-62621e3c5a71",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd /workspace/cfngle/raw-data/JM/raw-reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d7f1939d-faa8-4885-8605-0b7b624f3023",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "find -type f -exec md5sum \"{}\" + > checklist.chk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7a8233e4-6b90-499c-a60a-9bcc5c52bfd7",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./JM_031_metadata.txt: OK\n",
      "./JM_002.fastq: OK\n",
      "./JM_042_metadata.txt: OK\n",
      "./JM_027_metadata.txt: OK\n",
      "./JM_014.fastq: OK\n",
      "./JM_004.fastq: OK\n",
      "./JM_037_metadata.txt: OK\n",
      "./JM_042.fastq: OK\n",
      "./JM_008_metadata.txt: OK\n",
      "./.ipynb_checkpoints/SAMN21040240.fastq-checkpoint.txt: OK\n",
      "./.ipynb_checkpoints/SAMN21040242.fastq-checkpoint.txt: OK\n",
      "./.ipynb_checkpoints/filelist-checkpoint.txt: OK\n",
      "./.ipynb_checkpoints/SAMN21040241.fastq-checkpoint.txt: OK\n",
      "./JM_012_metadata.txt: OK\n",
      "./JM_033_metadata.txt: OK\n",
      "./JM_048_metadata.txt: OK\n",
      "./JM_025_metadata.txt: OK\n",
      "./JM_010_metadata.txt: OK\n",
      "./JM_017.fastq: OK\n",
      "./JM_039.fastq: OK\n",
      "./JM_045.fastq: OK\n",
      "./JM_019.fastq: OK\n",
      "./JM_018_metadata.txt: OK\n",
      "./JM_023.fastq: OK\n",
      "./JM_012.fastq: OK\n",
      "./JM_027.fastq: OK\n",
      "./JM_015.fastq: OK\n",
      "./JM_009_metadata.txt: OK\n",
      "./JM_033.fastq: OK\n",
      "./JM_045_metadata.txt: OK\n",
      "./JM_021_metadata.txt: OK\n",
      "./JM_029.fastq: OK\n",
      "./JM_026.fastq: OK\n",
      "./JM_001.fastq: OK\n",
      "./JM_039_metadata.txt: OK\n",
      "./JM_003.fastq: OK\n",
      "./JM_029_metadata.txt: OK\n",
      "./JM_031.fastq: OK\n",
      "./JM_011_metadata.txt: OK\n",
      "./JM_011.fastq: OK\n",
      "./JM_040.fastq: OK\n",
      "./JM_041.fastq: OK\n",
      "./JM_030_metadata.txt: OK\n",
      "./JM_019_metadata.txt: OK\n",
      "./JM_020.fastq: OK\n",
      "./JM_024_metadata.txt: OK\n",
      "./JM_022.fastq: OK\n",
      "./JM_047_metadata.txt: OK\n",
      "./JM_005.fastq: OK\n",
      "./JM_001_metadata.txt: OK\n",
      "./JM_026_metadata.txt: OK\n",
      "./JM_036.fastq: OK\n",
      "./JM_018.fastq: OK\n",
      "./JM_043.fastq: OK\n",
      "./JM_040_metadata.txt: OK\n",
      "./JM_038_metadata.txt: OK\n",
      "./JM_016_metadata.txt: OK\n",
      "./JM_024.fastq: OK\n",
      "./JM_017_metadata.txt: OK\n",
      "./JM_028.fastq: OK\n",
      "./JM_038.fastq: OK\n",
      "./JM_010.fastq: OK\n",
      "./JM_041_metadata.txt: OK\n",
      "./JM_035_metadata.txt: OK\n",
      "./JM_047.fastq: OK\n",
      "./JM_032.fastq: OK\n",
      "./JM_021.fastq: OK\n",
      "./checklist.chk: FAILED\n",
      "./JM_020_metadata.txt: OK\n",
      "./JM_023_metadata.txt: OK\n",
      "./JM_015_metadata.txt: OK\n",
      "./JM_002_metadata.txt: OK\n",
      "./JM_046.fastq: OK\n",
      "./JM_013.fastq: OK\n",
      "./JM_037.fastq: OK\n",
      "./JM_032_metadata.txt: OK\n",
      "./JM_044.fastq: OK\n",
      "./JM_025.fastq: OK\n",
      "./JM_028_metadata.txt: OK\n",
      "./JM_036_metadata.txt: OK\n",
      "./JM_014_metadata.txt: OK\n",
      "./JM_009.fastq: OK\n",
      "./JM_006.fastq: OK\n",
      "./JM_030.fastq: OK\n",
      "./JM_034_metadata.txt: OK\n",
      "./JM_006_metadata.txt: OK\n",
      "./JM_044_metadata.txt: OK\n",
      "./JM_016.fastq: OK\n",
      "./JM_013_metadata.txt: OK\n",
      "./JM_007_metadata.txt: OK\n",
      "./JM_046_metadata.txt: OK\n",
      "./JM_022_metadata.txt: OK\n",
      "./JM_003_metadata.txt: OK\n",
      "./JM_005_metadata.txt: OK\n",
      "./JM_043_metadata.txt: OK\n",
      "./JM_008.fastq: OK\n",
      "./JM_007.fastq: OK\n",
      "./JM_004_metadata.txt: OK\n",
      "./JM_034.fastq: OK\n",
      "./JM_035.fastq: OK\n",
      "md5sum: WARNING: 1 computed checksum did NOT match\n",
      "(/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "md5sum -c checklist.chk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e30aaf24-3fbe-4346-b0e3-4c197d377e52",
   "metadata": {},
   "source": [
    "### Define Project Variables and create analysis directories:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6c13cc36-5ae4-447e-9abe-32c08a774f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create analysis directories and define project variables\n",
    "\n",
    "# Define the user as a variable\n",
    "USER=\"cfngle\"\n",
    "\n",
    "# Define the project directory and temp subdirectory as a variable\n",
    "\n",
    "PROJECT=\"/workspace/$USER/raw-data/JM/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "810a60cc-673f-40fc-abe0-aab516ebcce4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Create analysis subdirectories\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM/\"\n",
    "mkdir -p $PROJECT\n",
    "mkdir -p $PROJECT/001.fastqc_raw\n",
    "mkdir -p $PROJECT/002.trim_galore\n",
    "mkdir -p $PROJECT/002.trimmomatic\n",
    "mkdir -p $PROJECT/003.fastq_trim_galore\n",
    "mkdir -p $PROJECT/004.alignments\n",
    "mkdir -p $PROJECT/004.alignments_tm\n",
    "mkdir -p $PROJECT/005.b.stats_alignments\n",
    "mkdir -p $PROJECT/005.qc_alignments\n",
    "mkdir -p $PROJECT/006.deduplication\n",
    "mkdir -p $PROJECT/007.meth_extraction\n",
    "mkdir -p $PROJECT/008.methylkit\n",
    "mkdir -p $PROJECT/099.test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5e17deb-bbbb-4bd7-910d-9d431cd550ee",
   "metadata": {},
   "source": [
    "### 1.1 FastQC RAW Data\n",
    "- The input for this step is the raw data from the provider in FASTQ format\n",
    "- The output from this step are the HTML FASTQC Reports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6563ed7d-f7b8-467a-9f26-f7b6617aa010",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./JM_002.fastq ./JM_014.fastq ./JM_004.fastq ./JM_042.fastq ./JM_017.fastq ./JM_039.fastq ./JM_045.fastq ./JM_019.fastq ./JM_023.fastq ./JM_012.fastq ./JM_027.fastq ./JM_015.fastq ./JM_033.fastq ./JM_029.fastq ./JM_026.fastq ./JM_001.fastq ./JM_003.fastq ./JM_031.fastq ./JM_011.fastq ./JM_040.fastq ./JM_041.fastq ./JM_020.fastq ./JM_022.fastq ./JM_005.fastq ./JM_036.fastq ./JM_018.fastq ./JM_043.fastq ./JM_024.fastq ./JM_028.fastq ./JM_038.fastq ./JM_010.fastq ./JM_047.fastq ./JM_032.fastq ./JM_021.fastq ./JM_046.fastq ./JM_013.fastq ./JM_037.fastq ./JM_044.fastq ./JM_025.fastq ./JM_009.fastq ./JM_006.fastq ./JM_030.fastq ./JM_016.fastq ./JM_008.fastq ./JM_007.fastq ./JM_034.fastq ./JM_035.fastq\n",
      "SBATCH_ARGS: --time 01:00:00 --mem 20G -c 10\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/001.fastqc_raw/fastqc_raw\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 47\n",
      "Submitted batch job 4454082\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\" #Define the location for the QC reports:\n",
    "IN=\"${PROJECT}/raw-reads\"\n",
    "OUT=\"${PROJECT}/001.fastqc_raw\"\n",
    "\n",
    "cd $IN\n",
    "#  FILES=`find -type f -name \"*.fastq\"`\n",
    "FILES=`find -type f -name \"*.fastq\"`\n",
    "echo $FILES\n",
    "\n",
    "\n",
    "# Load the FastQC module:\n",
    "module load FastQC\n",
    "\n",
    "for file in $FILES\n",
    "    do\n",
    "         COMMAND=\"fastqc --nogroup -q -t 2 -o ${OUT} ${file}\"\n",
    "         echo $COMMAND\n",
    "        #echo $file\n",
    "done | abatch -j ${OUT}/fastqc_raw --time 01:00:00 --mem 20G -c 10 | sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "cdccf7b3-24ba-4c00-ac81-f9c5d586a83c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /workspace/cfngle/raw-data/JM/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0908f545-c305-4531-b535-a3a61771e65e",
   "metadata": {},
   "outputs": [],
   "source": [
    "module load multiqc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "437edb0f-7d48-4a0e-9d2f-fcb49d2a2904",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=1708664591.5094242-395558;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.11\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mMultiQC Version v1.20 now available!\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report title: Fastqc-Raw\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /workspace/cfngle/raw-data/JM/001.fastqc_raw\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m97/97\u001b[0m  stqc_raw/JM_033_fastqc.html\u001b[0m.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 47 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : 001.fastqc_raw/Fastqc-Raw_multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : 001.fastqc_raw/Fastqc-Raw_multiqc_report_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n",
      "(/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "multiqc 001.fastqc_raw -o 001.fastqc_raw -i Fastqc-Raw"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50e04741-aff4-4af1-9279-0d6d60f29ae8",
   "metadata": {},
   "source": [
    "## 2. Trimming"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "325e34ec-e56e-4cbc-bed9-28b65bda686b",
   "metadata": {},
   "source": [
    "### 2.1a Trim_galore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5a155bfd-5e13-4f05-a07e-b71ffc78bfc4",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JM_001.fastq JM_002.fastq JM_003.fastq JM_004.fastq JM_005.fastq JM_006.fastq JM_007.fastq JM_008.fastq JM_009.fastq JM_010.fastq JM_011.fastq JM_012.fastq JM_013.fastq JM_014.fastq JM_015.fastq JM_016.fastq JM_017.fastq JM_018.fastq JM_019.fastq JM_020.fastq JM_021.fastq JM_022.fastq JM_023.fastq JM_024.fastq JM_025.fastq JM_026.fastq JM_027.fastq JM_028.fastq JM_029.fastq JM_030.fastq JM_031.fastq JM_032.fastq JM_033.fastq JM_034.fastq JM_035.fastq JM_036.fastq JM_037.fastq JM_038.fastq JM_039.fastq JM_040.fastq JM_041.fastq JM_042.fastq JM_043.fastq JM_044.fastq JM_045.fastq JM_046.fastq JM_047.fastq\n",
      "SBATCH_ARGS: --time 04:00:00 --mem 1G\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/002.trim_galore/trim_galore-logs\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 94\n",
      "Submitted batch job 4464148\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "IN=\"${PROJECT}/raw-reads\"\n",
    "OUT=\"${PROJECT}/002.trim_galore\"\n",
    "\n",
    "mkdir -p $IN\n",
    "mkdir -p $OUT\n",
    "\n",
    "\n",
    "# Get the files to trim:\n",
    "FILES=`basename -a ${IN}/*.fastq`\n",
    "\n",
    "echo $FILES\n",
    "\n",
    "module load trim_galore\n",
    "\n",
    "for FILE in $FILES\n",
    "     do\n",
    "        In_File1=${IN}/${FILE}\n",
    "        echo $In_File1\n",
    "        Out=${OUT}/${FILE}_trimmed.fastq\n",
    "        COMMAND=\"trim_galore --rrbs --quality 25 --clip_R1 4 --three_prime_clip_R1 2 $In_File1 --output_dir $OUT\"\n",
    "        \n",
    "        echo $COMMAND\n",
    "done | abatch -j ${OUT}/trim_galore-logs --time 04:00:00 --mem 1G | sbatch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cab78171-b1c8-4bc8-894d-cc73252af1bf",
   "metadata": {},
   "source": [
    "### 2.1b Trimmomatic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b972d276-6313-4f3b-8e0b-4e0c19f53a75",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SBATCH_ARGS: --time 04:00:00 --mem 1G\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/002.trimmomatic//trimmomatic-logs\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 94\n",
      "Submitted batch job 4468474\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "IN=\"${PROJECT}/raw-reads\"\n",
    "OUT=\"${PROJECT}/002.trimmomatic/\"\n",
    "ADAPTER=\"${PROJECT}/zzz.adapters/TruSeq3-SE.fa\"\n",
    "\n",
    "mkdir -p $IN\n",
    "mkdir -p $OUT\n",
    "\n",
    "\n",
    "# Get the files to trim:\n",
    "FILES=`basename -a ${IN}/*.fastq`\n",
    "\n",
    "#echo $FILES\n",
    "\n",
    "module load Trimmomatic\n",
    "\n",
    "for FILE in $FILES\n",
    "     do\n",
    "        In_File1=${IN}/${FILE}\n",
    "        echo $In_File1\n",
    "        Out=${OUT}${FILE}_trimmoniac.fastq\n",
    "        COMMAND=\"trimmomatic SE -phred33 ${In_File1} ${Out} ILLUMINACLIP:${ADAPTER}:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\"\n",
    "        \n",
    "        echo $COMMAND\n",
    "done | abatch -j ${OUT}/trimmomatic-logs --time 04:00:00 --mem 1G | sbatch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f93e76c-40f1-4c3b-9466-c6eaa788cd00",
   "metadata": {},
   "source": [
    "### 2.2a FastQC TRIMMED Data\n",
    "- The input for this step is the raw data from the provider in FASTQ format\n",
    "- The output from this step are the HTML FASTQC Reports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "51ce841f-d25c-452a-be07-537f33193bbc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading \u001b[1mFastQC/0.11.7\u001b[22m\n",
      "  \u001b[91mERROR\u001b[0m: Conflicting 'FastQC' is loaded\n",
      "SBATCH_ARGS: --time 02:00:00 --mem 1G\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/003.fastq_trim_galore/fastqc-trim_galore-logs\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 47\n",
      "Submitted batch job 4465018\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\" # Define the location for the QC reports:\n",
    "# Define the location for the QC reports:\n",
    "IN=\"${PROJECT}/002.trim_galore\"\n",
    "OUT=\"${PROJECT}/003.fastq_trim_galore\"\n",
    "\n",
    "# Define the list of files to process:\n",
    "FILES=`ls ${IN}/*.fq`\n",
    "\n",
    "# Load the FastQC module:\n",
    "module load FastQC\n",
    "\n",
    "for file in $FILES\n",
    "    do\n",
    "        COMMAND=\"fastqc --nogroup -q -t 2 -o ${OUT} ${file}\"\n",
    "        echo $COMMAND\n",
    "done | abatch -j ${OUT}/fastqc-trim_galore-logs --time 02:00:00 --mem 1G | sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "165399e1-0c53-4375-9f83-ef419f9a256f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** INFO ** : singularity has been deprecated - please use apptainer in place.\n",
      "\n",
      "Loading \u001b[1mmultiqc/1.11\u001b[22m\n",
      "  \u001b[94mLoading requirement\u001b[0m: singularity/3.10.3\n"
     ]
    }
   ],
   "source": [
    "module load multiqc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "aeae0461-94aa-48ef-bc35-44e85d8242ac",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=1709088715.1051173-869333;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.11\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mMultiQC Version v1.20 now available!\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report title: Fastqc-Trimmed\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /workspace/cfngle/raw-data/JM/003.fastq_trim_galore\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m200/200\u001b[0m  0m[0m  \n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 47 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : 003.fastq_trim_galore/Fastqc-Trimmed_multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : 003.fastq_trim_galore/Fastqc-Trimmed_multiqc_report_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "cd /workspace/cfngle/raw-data/JM\n",
    "multiqc 003.fastq_trim_galore -o 003.fastq_trim_galore -i Fastqc-Trimmed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "030cba67-0c6d-41ea-b99d-ead2110c7803",
   "metadata": {},
   "source": [
    "### 2.2b FastQC TRIMMED Data\n",
    "- The input for this step is the raw data from the provider in FASTQ format\n",
    "- The output from this step are the HTML FASTQC Reports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "51794245-fa9d-4b40-b403-2b78f22072a6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading \u001b[1mFastQC/0.11.7\u001b[22m\n",
      "  \u001b[91mERROR\u001b[0m: Conflicting 'FastQC' is loaded\n",
      "SBATCH_ARGS: --time 02:00:00 --mem 1G\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/003.fastq_trimmomatic/fastqc-trimmomatic-logs\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 47\n",
      "Submitted batch job 4468568\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\" # Define the location for the QC reports:\n",
    "# Define the location for the QC reports:\n",
    "IN=\"${PROJECT}/002.trimmomatic\"\n",
    "OUT=\"${PROJECT}/003.fastq_trimmomatic\"\n",
    "\n",
    "# Define the list of files to process:\n",
    "FILES=`ls ${IN}/*.fastq`\n",
    "\n",
    "# Load the FastQC module:\n",
    "module load FastQC\n",
    "\n",
    "for file in $FILES\n",
    "    do\n",
    "        COMMAND=\"fastqc --nogroup -q -t 2 -o ${OUT} ${file}\"\n",
    "        echo $COMMAND\n",
    "done | abatch -j ${OUT}/fastqc-trimmomatic-logs --time 02:00:00 --mem 1G | sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "22a2abf7-c41f-4f69-9789-f26d4941b705",
   "metadata": {},
   "outputs": [],
   "source": [
    "module load multiqc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "339514d2-0400-415b-bcb7-3825b08caa69",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=1709164268.7307322-895457;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.11\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mMultiQC Version v1.21 now available!\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report title: Fastqc-Trimmed\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /workspace/cfngle/raw-data/JM/003.fastq_trimmomatic\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m189/189\u001b[0m  0m[0m  \n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 47 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : 003.fastq_trimmomatic/Fastqc-Trimmed_multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : 003.fastq_trimmomatic/Fastqc-Trimmed_multiqc_report_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "cd /workspace/cfngle/raw-data/JM\n",
    "multiqc 003.fastq_trimmomatic -o 003.fastq_trimmomatic -i Fastqc-Trimmed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "105a8a9b-d575-4ea9-8651-ab1d2eedc6eb",
   "metadata": {},
   "source": [
    "## 5. Alignment\n",
    "Bismark user guide https://felixkrueger.github.io/Bismark/\n",
    "### 5.1 Preparation of the genome\n",
    "Japanese medaka rgenome v1 (ASM223467v1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b0d6738f-fcaf-4e38-99f3-71603447c489",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd rgenome/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e7e5d3e0-1ae5-4b58-b0de-0a9686bbb058",
   "metadata": {},
   "outputs": [],
   "source": [
    "grep -e \">\" GCF_002234675.1_ASM223467v1_genomic.fna | awk 'sub(/^>/, \"\")' | sort -k1 > JM_chromosomes.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "dfb354d0-c0c6-42c8-8fe2-60d1256d78ac",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_004387.1 Oryzias latipes mitochondrion, complete genome\n",
      "NC_019859.2 Oryzias latipes chromosome 1, ASM223467v1\n",
      "NC_019860.2 Oryzias latipes chromosome 2, ASM223467v1\n",
      "NC_019861.2 Oryzias latipes chromosome 3, ASM223467v1\n",
      "NC_019862.2 Oryzias latipes chromosome 4, ASM223467v1\n",
      "NC_019863.2 Oryzias latipes chromosome 5, ASM223467v1\n",
      "NC_019864.2 Oryzias latipes chromosome 6, ASM223467v1\n",
      "NC_019865.2 Oryzias latipes chromosome 7, ASM223467v1\n",
      "NC_019866.2 Oryzias latipes chromosome 8, ASM223467v1\n",
      "NC_019867.2 Oryzias latipes chromosome 9, ASM223467v1\n"
     ]
    }
   ],
   "source": [
    "head JM_chromosomes.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "31bdff52-c31e-4e7e-8937-44177045ed93",
   "metadata": {},
   "outputs": [],
   "source": [
    "module load bismark/0.23.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "fb0d2214-3c91-4125-a40d-51d0456ee89d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing bisulfite genomes out into a single MFA (multi FastA) file\n",
      "\n",
      "Bisulfite Genome Indexer version v0.23.0 (last modified: 14 April 2019)\n",
      "\n",
      "Step I - Prepare genome folders - completed\n",
      "\n",
      "\n",
      "Calculating genomic frequencies (this may take several minutes depending on genome size) ...\n",
      "====================================================================================================================================================================\n",
      "Now reading in and storing sequence information of the genome specified in: /workspace/cfngle/raw-data/JM/rgenome/\n",
      "\n",
      "chr NC_019859.2 (37713152 bp)\n",
      "chr NC_019860.2 (25379070 bp)\n",
      "chr NC_019861.2 (38248663 bp)\n",
      "chr NC_019862.2 (32868862 bp)\n",
      "chr NC_019863.2 (33205099 bp)\n",
      "chr NC_019864.2 (32246747 bp)\n",
      "chr NC_019865.2 (34573382 bp)\n",
      "chr NC_019866.2 (26239357 bp)\n",
      "chr NC_019867.2 (33399407 bp)\n",
      "chr NC_019868.2 (31218526 bp)\n",
      "chr NC_019869.2 (28210532 bp)\n",
      "chr NC_019870.2 (30543476 bp)\n",
      "chr NC_019871.2 (33825776 bp)\n",
      "chr NC_019872.2 (30598983 bp)\n",
      "chr NC_019873.2 (30476034 bp)\n",
      "chr NC_019874.2 (32958677 bp)\n",
      "chr NC_019875.2 (31792230 bp)\n",
      "chr NC_019876.2 (30918796 bp)\n",
      "chr NC_019877.2 (25472880 bp)\n",
      "chr NC_019878.2 (25942153 bp)\n",
      "chr NC_019879.2 (31148813 bp)\n",
      "chr NC_019880.2 (28976614 bp)\n",
      "chr NC_019881.2 (24400806 bp)\n",
      "chr NC_019882.2 (23682337 bp)\n",
      "chr NC_004387.1 (16714 bp)\n",
      "\n",
      "Processing chromosome >> NC_019869.2 <<\n",
      "Processing chromosome >> NC_019864.2 <<\n",
      "Processing chromosome >> NC_019861.2 <<\n",
      "Processing chromosome >> NC_019866.2 <<\n",
      "Processing chromosome >> NC_019879.2 <<\n",
      "Processing chromosome >> NC_019874.2 <<\n",
      "Processing chromosome >> NC_004387.1 <<\n",
      "Processing chromosome >> NC_019871.2 <<\n",
      "Processing chromosome >> NC_019881.2 <<\n",
      "Processing chromosome >> NC_019876.2 <<\n",
      "Processing chromosome >> NC_019877.2 <<\n",
      "Processing chromosome >> NC_019873.2 <<\n",
      "Processing chromosome >> NC_019865.2 <<\n",
      "Processing chromosome >> NC_019882.2 <<\n",
      "Processing chromosome >> NC_019880.2 <<\n",
      "Processing chromosome >> NC_019870.2 <<\n",
      "Processing chromosome >> NC_019859.2 <<\n",
      "Processing chromosome >> NC_019878.2 <<\n",
      "Processing chromosome >> NC_019872.2 <<\n",
      "Processing chromosome >> NC_019867.2 <<\n",
      "Processing chromosome >> NC_019863.2 <<\n",
      "Processing chromosome >> NC_019875.2 <<\n",
      "Processing chromosome >> NC_019860.2 <<\n",
      "Processing chromosome >> NC_019868.2 <<\n",
      "Processing chromosome >> NC_019862.2 <<\n",
      "Writing genomic nucleotide frequencies to the file >/workspace/cfngle/raw-data/JM/rgenome/genomic_nucleotide_frequencies.txt< for future re-use\n",
      "Writing count of (di-)nucleotide: A\t216957530\n",
      "Writing count of (di-)nucleotide: AA\t76267116\n",
      "Writing count of (di-)nucleotide: AC\t41125942\n",
      "Writing count of (di-)nucleotide: AG\t46339810\n",
      "Writing count of (di-)nucleotide: AT\t53224515\n",
      "Writing count of (di-)nucleotide: C\t149808184\n",
      "Writing count of (di-)nucleotide: CA\t54903461\n",
      "Writing count of (di-)nucleotide: CC\t33648602\n",
      "Writing count of (di-)nucleotide: CG\t14873895\n",
      "Writing count of (di-)nucleotide: CT\t46382122\n",
      "Writing count of (di-)nucleotide: G\t149806542\n",
      "Writing count of (di-)nucleotide: GA\t43510171\n",
      "Writing count of (di-)nucleotide: GC\t31502622\n",
      "Writing count of (di-)nucleotide: GG\t33674420\n",
      "Writing count of (di-)nucleotide: GT\t41119214\n",
      "Writing count of (di-)nucleotide: T\t216993830\n",
      "Writing count of (di-)nucleotide: TA\t42276625\n",
      "Writing count of (di-)nucleotide: TC\t43530907\n",
      "Writing count of (di-)nucleotide: TG\t54918326\n",
      "Writing count of (di-)nucleotide: TT\t76267822\n",
      "Finished processing genomic nucleotide frequencies\n",
      "\n",
      "\n",
      "Total number of conversions performed:\n",
      "C->T:\t149808184\n",
      "G->A:\t149806542\n",
      "\n",
      "Step II - Genome bisulfite conversions - completed\n",
      "\n",
      "\n",
      "Bismark Genome Preparation - Step III: Launching the Bowtie 2 indexer\n",
      "Please be aware that this process can - depending on genome size - take several hours!\n",
      "Settings:\n",
      "  Output files: \"BS_CT.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genome_mfa.CT_conversion.fa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "Settings:\n",
      "  Output files: \"BS_GA.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genome_mfa.GA_conversion.fa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:06\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time reading reference sizes: 00:00:06\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:03\n",
      "bmax according to bmaxDivN setting: 183391521\n",
      "Using parameters --bmax 137543641 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Time to join reference sequences: 00:00:03\n",
      "bmax according to bmaxDivN setting: 183391521\n",
      "Using parameters --bmax 137543641 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 137543641 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Passed!  Constructing with these parameters: --bmax 137543641 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:20\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  V-Sorting samples time: 00:00:20\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:05\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Ranking v-sort output time: 00:00:05\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:08\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 7.33566e+08 (target: 137543640)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:09\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 7.33566e+08 (target: 137543640)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 733566086 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block of length 733566086 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:08:08\n",
      "Returning block of 733566087 for bucket 1\n",
      "  Sorting block time: 00:09:56\n",
      "Returning block of 733566087 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 366764072\n",
      "fchr[G]: 516572256\n",
      "fchr[T]: 516572256\n",
      "fchr[$]: 733566086\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 248723414 bytes to primary EBWT file: BS_GA.1.bt2\n",
      "Wrote 183391528 bytes to secondary EBWT file: BS_GA.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 733566086\n",
      "    bwtLen: 733566087\n",
      "    sz: 183391522\n",
      "    bwtSz: 183391522\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 45847881\n",
      "    offsSz: 183391524\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 3820657\n",
      "    numLines: 3820657\n",
      "    ebwtTotLen: 244522048\n",
      "    ebwtTotSz: 244522048\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:10:47\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:03\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:03\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 183391521\n",
      "Using parameters --bmax 137543641 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 137543641 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:20\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:05\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:08\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 7.33566e+08 (target: 137543640)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 733566086 for bucket 1\n",
      "  (Using difference cover)\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 216957530\n",
      "fchr[G]: 216957530\n",
      "fchr[T]: 366764072\n",
      "fchr[$]: 733566086\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 248723414 bytes to primary EBWT file: BS_CT.1.bt2\n",
      "Wrote 183391528 bytes to secondary EBWT file: BS_CT.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 733566086\n",
      "    bwtLen: 733566087\n",
      "    sz: 183391522\n",
      "    bwtSz: 183391522\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 45847881\n",
      "    offsSz: 183391524\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 3820657\n",
      "    numLines: 3820657\n",
      "    ebwtTotLen: 244522048\n",
      "    ebwtTotSz: 244522048\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:12:34\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:03\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:03\n",
      "  Time to reverse reference sequence: 00:00:01\n",
      "bmax according to bmaxDivN setting: 183391521\n",
      "Using parameters --bmax 137543641 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 137543641 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:20\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:05\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:09\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 7.33566e+08 (target: 137543640)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 733566086 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:08:18\n",
      "Returning block of 733566087 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 366764072\n",
      "fchr[G]: 516572256\n",
      "fchr[T]: 516572256\n",
      "fchr[$]: 733566086\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 248723414 bytes to primary EBWT file: BS_GA.rev.1.bt2\n",
      "Wrote 183391528 bytes to secondary EBWT file: BS_GA.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 733566086\n",
      "    bwtLen: 733566087\n",
      "    sz: 183391522\n",
      "    bwtSz: 183391522\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 45847881\n",
      "    offsSz: 183391524\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 3820657\n",
      "    numLines: 3820657\n",
      "    ebwtTotLen: 244522048\n",
      "    ebwtTotSz: 244522048\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:10:54\n",
      "  Sorting block time: 00:10:15\n",
      "Returning block of 733566087 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 216957530\n",
      "fchr[G]: 216957530\n",
      "fchr[T]: 366764072\n",
      "fchr[$]: 733566086\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 248723414 bytes to primary EBWT file: BS_CT.rev.1.bt2\n",
      "Wrote 183391528 bytes to secondary EBWT file: BS_CT.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 733566086\n",
      "    bwtLen: 733566087\n",
      "    sz: 183391522\n",
      "    bwtSz: 183391522\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 45847881\n",
      "    offsSz: 183391524\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 3820657\n",
      "    numLines: 3820657\n",
      "    ebwtTotLen: 244522048\n",
      "    ebwtTotSz: 244522048\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:12:48\n"
     ]
    }
   ],
   "source": [
    "bismark_genome_preparation --genomic_composition /workspace/cfngle/raw-data/JM/rgenome/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c9fecb5-d184-4c1d-a649-f10c6badd0e1",
   "metadata": {},
   "source": [
    "### 5.2 Alignments to genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0f373877-37ea-4284-8322-93f1233f1e50",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SBATCH_ARGS: --time 4-24:00:00 --mem 80G -c 5\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/004.alignments/bis-align\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 47\n",
      "Submitted batch job 4465104\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "IN=\"${PROJECT}/002.trim_galore\"\n",
    "OUT=\"${PROJECT}/004.alignments\"\n",
    "GENOME=\"/workspace/cfngle/raw-data/JM/rgenome\"\n",
    "LOG=\"${OUT}/logs\"\n",
    "\n",
    "mkdir -p $LOG\n",
    "\n",
    "module load bismark/0.23.0\n",
    "\n",
    "# Use echo statements to be sure that the results from awk are what you really want...\n",
    "# FILES=`ls ${IN}/*.fq | awk -F'[_ ]' '{print $1\"_\"$2\"_\"$3\"_\"$4\"_\"$5}' | sort -u`\n",
    "FILES=`ls ${IN}/*.fq`\n",
    "# echo $FILES\n",
    "\n",
    "for file in ${FILES}\n",
    "    do\n",
    "        NAME=`basename ${file}`\n",
    "        PREFIX=`echo ${NAME} | awk -F[_] '{print $1\"_\"$2\"_Alignments\"}'`\n",
    "        #echo ${PREFIX}\n",
    "        #echo ${file}\n",
    "        COMMAND=\"bismark \\\n",
    "                --genome ${GENOME} \\\n",
    "                $file \\\n",
    "                --o ${OUT}/${PREFIX} \\\n",
    "                --multicore 5 \\\n",
    "                --non_directional\"\n",
    "       # echo \"${OUT}${PREFIX}\"\n",
    "       echo $COMMAND\n",
    "\n",
    "done | abatch -j ${OUT}/bis-align-local --time 23:00:00 --mem 25G -c 1 | sbatch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e444f8a9-ffb0-4e86-b4a2-e96129533350",
   "metadata": {},
   "source": [
    "#### 5.2.1 Alignments to genome trim_galore BISMARK --local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "2a4f44ac-1fd2-4005-90dc-0c88f2d9e05e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SBATCH_ARGS: --time 2-23:59:00 --mem 25G -c 1\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/004.alignments_BS_local/bis-align_local\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 47\n",
      "Submitted batch job 4676246\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "IN=\"${PROJECT}/002.trim_galore\"\n",
    "OUT=\"${PROJECT}/004.alignments_BM_local\"\n",
    "GENOME=\"/workspace/cfngle/raw-data/JM/rgenome\"\n",
    "LOG=\"${OUT}/logs\"\n",
    "\n",
    "mkdir -p $LOG\n",
    "\n",
    "module load bismark/0.23.0\n",
    "\n",
    "# Use echo statements to be sure that the results from awk are what you really want...\n",
    "# FILES=`ls ${IN}/*.fq | awk -F'[_ ]' '{print $1\"_\"$2\"_\"$3\"_\"$4\"_\"$5}' | sort -u`\n",
    "FILES=`ls ${IN}/*.fq`\n",
    "# echo $FILES\n",
    "\n",
    "for file in ${FILES}\n",
    "    do\n",
    "        NAME=`basename ${file}`\n",
    "        PREFIX=`echo ${NAME} | awk -F[_] '{print $1\"_\"$2\"_Alignments\"}'`\n",
    "        #echo ${PREFIX}\n",
    "        #echo ${file}\n",
    "        COMMAND=\"bismark \\\n",
    "                --genome ${GENOME} \\\n",
    "                $file \\\n",
    "                --o ${OUT}/${PREFIX} \\\n",
    "                --multicore 5 \\\n",
    "                --non_directional \\\n",
    "                --local\"\n",
    "       # echo \"${OUT}${PREFIX}\"\n",
    "       echo $COMMAND\n",
    "\n",
    "done | abatch -j ${OUT}/bis-align_local --time 2-23:59:00 --mem 25G -c 1 | sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "ecdc372d-e412-4323-aae6-2b8a1c9cbe7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=1710210739.8987062-991077;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.11\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mMultiQC Version v1.21 now available!\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report title: QC.alignments\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /workspace/cfngle/raw-data/JM/004.alignments_BS_local\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m190/190\u001b[0m  0m[0m  \n",
      "\u001b[?25h\u001b[34m|           bismark\u001b[0m | Found 47 alignment reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : 005.qc_alignments_BS_local/QC.alignments_multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : 005.qc_alignments_BS_local/QC.alignments_multiqc_report_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "multiqc 004.alignments_BM_local -i QC.alignments -o 005.qc_alignments_BM_local"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "467de58a-18ff-484d-a3ae-166efa68e6ed",
   "metadata": {},
   "source": [
    "#### 5.2.1 Alignments to genome (Trimmomatic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "611a95cf-1b25-4632-a7db-556efecb462a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SBATCH_ARGS: --time 4-24:00:00 --mem 80G -c 8\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/004.alignments_tm/bis-align\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 47\n",
      "Submitted batch job 4483607\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "IN=\"${PROJECT}/002.trimmomatic\"\n",
    "OUT=\"${PROJECT}/004.alignments_tm\"\n",
    "GENOME=\"/workspace/cfngle/raw-data/JM/rgenome\"\n",
    "LOG=\"${OUT}/logs\"\n",
    "\n",
    "mkdir -p $LOG\n",
    "\n",
    "module load bismark/0.23.0\n",
    "\n",
    "# Use echo statements to be sure that the results from awk are what you really want...\n",
    "# FILES=`ls ${IN}/*.fq | awk -F'[_ ]' '{print $1\"_\"$2\"_\"$3\"_\"$4\"_\"$5}' | sort -u`\n",
    "FILES=`ls ${IN}/*.fastq`\n",
    "# echo $FILES\n",
    "\n",
    "for file in ${FILES}\n",
    "    do\n",
    "        NAME=`basename ${file}`\n",
    "        PREFIX=`echo ${NAME} | awk -F[_] '{print $1\"_\"$2\"_Alignments\"}'`\n",
    "        #echo ${PREFIX}\n",
    "        #echo ${file}\n",
    "        COMMAND=\"bismark \\\n",
    "                --genome ${GENOME} \\\n",
    "                $file \\\n",
    "                --o ${OUT}/${PREFIX} \\\n",
    "                --multicore 4 \\\n",
    "                --non_directional\"\n",
    "       # echo \"${OUT}${PREFIX}\"\n",
    "       echo $COMMAND\n",
    "\n",
    "done | abatch -j ${OUT}/bis-align --time 4-24:00:00 --mem 10G -c 1| sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "0e0dd3e6-e2fa-43a3-a012-11796117d48e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=1709671700.8706174-644690;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.11\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mMultiQC Version v1.21 now available!\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report title: QC.alignments\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /workspace/cfngle/raw-data/JM/004.alignments_tm\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m191/191\u001b[0m  0m[0m  \n",
      "\u001b[?25h\u001b[34m|           bismark\u001b[0m | Found 47 alignment reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : 005.qc.alignments_tm/QC.alignments_multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : 005.qc.alignments_tm/QC.alignments_multiqc_report_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "multiqc 004.alignments_tm -i QC.alignments -o 005.qc.alignments_tm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e6b06c2-ad91-431d-b625-e8efa8219abd",
   "metadata": {},
   "source": [
    "### 5.1a BSseeker"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3970845-a09e-4153-9bd7-c3a50a1dfad8",
   "metadata": {},
   "source": [
    "##### BSseeker did not work with command line directly. A bioconda environment had to be set up and BSseeker 2 directly installed there. (conda install bioconda/label/cf201901::bs-seeker2).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd839b4d-dc2a-43af-81ab-d9584c9a3450",
   "metadata": {},
   "source": [
    "#### Script for BS indexing (for .sh file) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ebeae49-1865-4ffd-b132-2bd3046d0128",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --job-name=indexBS\n",
    "#SBATCH --time=24:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/workspace/cfngle/raw-data/JM/rgenome/output.log\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "GENOME=\"/workspace/cfngle/raw-data/JM/rgenome\"\n",
    "OUT=$GENOME\n",
    "\n",
    "module load conda\n",
    "conda deactivate\n",
    "conda activate cfngle_env-01\n",
    "\n",
    "bs_seeker2-build.py \\\n",
    "    -f ${GENOME}/GCF_002234675.1_ASM223467v1_genomic.fasta \\\n",
    "    --aligner bowtie2 \\\n",
    "    -r \\\n",
    "    -d /workspace/cfngle/raw-data/JM/rgenome/BSseeker2-index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d895d3db-a3a2-4050-beb4-7763e24eb63c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 23.1.0\n",
      "  latest version: 24.1.2\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=24.1.2\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n",
      "(/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda install bowtie2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b1e9b088-37a7-4f0b-8af5-f511b863bae6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: bs_seeker2-build.py [options]\n",
      "\n",
      "Options:\n",
      "  -h, --help            show this help message and exit\n",
      "  -f FILE, --file=FILE  Input your reference genome file (fasta)\n",
      "  --aligner=ALIGNER     Aligner program to perform the analysis: bowtie,\n",
      "                        bowtie2, soap, rmap [Default: bowtie]\n",
      "  -p PATH, --path=PATH  Path to the aligner program. Detected:\n",
      "                        bowtie: None\n",
      "                        bowtie2:\n",
      "                        /workspace/appscratch/miniconda/cfngle_env-01/bin\n",
      "                        rmap: None\n",
      "                        soap: None\n",
      "  -d DBPATH, --db=DBPATH\n",
      "                        Path to the reference genome library (generated in\n",
      "                        preprocessing genome) [Default: /powerplant/workspace/\n",
      "                        appscratch/miniconda/cfngle_env-01/bin/bs_utils/refere\n",
      "                        nce_genomes]\n",
      "  -v, --version         show version of BS-Seeker2\n",
      "\n",
      "  Reduced Representation Bisulfite Sequencing Options:\n",
      "    Use this options with conjuction of -r [--rrbs]\n",
      "\n",
      "    -r, --rrbs          Build index specially for Reduced Representation\n",
      "                        Bisulfite Sequencing experiments. Genome other than\n",
      "                        certain fragments will be masked. [Default: False]\n",
      "    -l LOW_BOUND, --low=LOW_BOUND\n",
      "                        lower bound of fragment length (excluding recognition\n",
      "                        sequence such as C-CGG) [Default: 20]\n",
      "    -u UP_BOUND, --up=UP_BOUND\n",
      "                        upper bound of fragment length (excluding recognition\n",
      "                        sequence such as C-CGG ends) [Default: 500]\n",
      "    -c CUT_FORMAT, --cut-site=CUT_FORMAT\n",
      "                        Cut sites of restriction enzyme. Ex: MspI(C-CGG),\n",
      "                        Mael:(C-TAG), double-enzyme MspI&Mael:(C-CGG,C-TAG).\n",
      "                        [Default: C-CGG]\n",
      "(/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "bs_seeker2-build.py -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6352cee-77ed-4a3d-8129-6b464c28747e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfngle_env-01) \n",
      "     BS-Seeker2 v2.1.7 - Oct. 23, 2018\n",
      "\n",
      "Reference genome file: /workspace/cfngle/raw-data/ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fasta\n",
      "Reduced Representation Bisulfite Sequencing: True\n",
      "Short reads aligner you are using: bowtie2\n",
      "Builder path: /workspace/appscratch/miniconda/cfngle_env-01/bin/bowtie2-build\n"
     ]
    }
   ],
   "source": [
    "# indexing reference genome for BSseeker\n",
    "bs_seeker2-build.py -f /workspace/cfngle/raw-data/ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fasta --aligner bowtie2 -r -d /workspace/cfngle/raw-data/ZF/rgenome/BSseeker2-index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "5ddcbb9f-2da3-43b1-a617-1066a35928f4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) SBATCH_ARGS: --time 4-24:00:00 --mem 80G -c 8\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/004.alignments_BS_tm/BS-align\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 47\n",
      "Submitted batch job 4484673\n",
      "(/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "IN=\"${PROJECT}/002.trim_galore\"\n",
    "OUT=\"${PROJECT}/004.alignments_BS_tm\"\n",
    "GENOME=\"/workspace/cfngle/raw-data/JM/rgenome\"\n",
    "LOG=\"${OUT}/logs\"\n",
    "\n",
    "mkdir -p $LOG\n",
    "\n",
    "module load conda\n",
    "conda deactivate\n",
    "conda activate cfngle_env-01\n",
    "\n",
    "# Use echo statements to be sure that the results from awk are what you really want...\n",
    "# FILES=`ls ${IN}/*.fq | awk -F'[_ ]' '{print $1\"_\"$2\"_\"$3\"_\"$4\"_\"$5}' | sort -u`\n",
    "FILES=`ls ${IN}/*.fq`\n",
    "#echo $FILES\n",
    "\n",
    "for file in ${FILES}\n",
    "    do\n",
    "        NAME=`basename ${file}`\n",
    "        PREFIX=`echo ${NAME} | awk -F[_] '{print $1\"_\"$2\"_Alignments_BS_tm\"}'`\n",
    "        #echo ${PREFIX}\n",
    "        #echo ${IN}/${NAME}\n",
    "        COMMAND=\"bs_seeker2-align.py \\\n",
    "            -i ${IN}/$NAME \\\n",
    "            -r \\\n",
    "            --aligner bowtie2 \\\n",
    "            -o ${OUT}/${PREFIX}.bam \\\n",
    "            -g ${GENOME}/GCF_002234675.1_ASM223467v1_genomic.fasta \\\n",
    "            -d ${GENOME}/BSseeker2-index \\\n",
    "            --temp_dir=${OUT}/TMP/ \\\n",
    "            --bt2-p 4\"\n",
    "       echo $COMMAND\n",
    "\n",
    "done | abatch -j ${OUT}/BS-align --time 24:00:00 --mem 10G -c 1| sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "a2b29e12-8dda-4404-b5df-aae013e70a3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda deactivate\n",
    "cd /workspace/cfngle/raw-data/JM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b9aabc7b-1fb3-4d13-a010-4fb159ea9b58",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading \u001b[1mFastQC/0.11.7\u001b[22m\n",
      "  \u001b[91mERROR\u001b[0m: Conflicting 'FastQC' is loaded\n",
      "SBATCH_ARGS: --time 02:00:00 --mem 1G\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/005.fastq_BS_tm/fastqc-BS_tm-logs\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 47\n",
      "Submitted batch job 4483213\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\" # Define the location for the QC reports:\n",
    "# Define the location for the QC reports:\n",
    "IN=\"${PROJECT}/004.alignments_BS_tm\"\n",
    "OUT=\"${PROJECT}/005.fastq_BS_tm\"\n",
    "\n",
    "# Define the list of files to process:\n",
    "FILES=`ls ${IN}/*.bam`\n",
    "\n",
    "# Load the FastQC module:\n",
    "module load FastQC\n",
    "\n",
    "for file in $FILES\n",
    "    do\n",
    "        COMMAND=\"fastqc --nogroup -q -t 2 -o ${OUT} ${file}\"\n",
    "        echo $COMMAND\n",
    "done | abatch -j ${OUT}/fastqc-BS_tm-logs --time 02:00:00 --mem 1G | sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "196e27c0-8347-44de-bc68-607daf9c15f1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=1709586683.9206378-841825;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.11\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mMultiQC Version v1.21 now available!\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report title: QC.alignments\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /workspace/cfngle/raw-data/JM/005.fastq_BS_tm\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m189/189\u001b[0m  0m[0m  \n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 47 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : 005.fastq_BS_tm/QC.alignments_multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : 005.fastq_BS_tm/QC.alignments_multiqc_report_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "multiqc 005.fastq_BS_tm -i QC.alignments -o 005.fastq_BS_tm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "8416d974-3e2b-489c-8763-464589e6a139",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) (/workspace/appscratch/miniconda/cfngle_env-01) SBATCH_ARGS: --time 24:00:00 --mem 20 -c 2\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/007.meth_extraction_BS_tm/BS_extract\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 47\n",
      "Submitted batch job 4505706\n",
      "(/workspace/appscratch/miniconda/cfngle_env-01) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "IN=\"${PROJECT}/004.alignments_BS_tm\"\n",
    "OUT=\"${PROJECT}/007.meth_extraction_BS_tm\"\n",
    "LOG=\"${OUT}/logs\"\n",
    "\n",
    "mkdir -p $LOG\n",
    "\n",
    "module load conda\n",
    "conda deactivate\n",
    "conda activate cfngle_env-01\n",
    "\n",
    "FILES=`ls ${IN}/*.bam`\n",
    "#echo $FILES\n",
    "for file in ${FILES}\n",
    "    do\n",
    "        NAME=`basename ${file}`\n",
    "        PREFIX=`echo ${NAME} | awk -F[_] '{print $1\"_\"$2\"_meth_extract_BS_tm\"}'`\n",
    "        #echo $NAME\n",
    "        COMMAND=\"bs_seeker2-call_methylation.py \\\n",
    "            -i ${IN}/$NAME \\\n",
    "            -o ${OUT}/${PREFIX} \\\n",
    "            -d ${GENOME}/BSseeker2-index/GCF_002234675.1_ASM223467v1_genomic.fasta_rrbs_20_500_bowtie2/\"\n",
    "        echo $COMMAND\n",
    "\n",
    "done | abatch -j ${OUT}/BS_extract --time 24:00:00 --mem 8G -c 1 | sbatch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1be44dfc-32ae-4e64-9145-0844b8093894",
   "metadata": {},
   "source": [
    "### 7. Methylation extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "765ab234-84b0-440b-b723-566a2b9a0836",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SBATCH_ARGS: --time 5-01:00:00 --mem 20G -c 6\n",
      "JOB_ARRAY_NAME: /workspace/cfngle/raw-data/JM/007.meth_extraction/meth-extract\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 47\n",
      "Submitted batch job 4468888\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "IN=\"${PROJECT}/004.alignments\"\n",
    "OUT=\"${PROJECT}/007.meth_extraction\"\n",
    "GENOME=\"/workspace/cfngle/raw-data/JM/rgenome\"\n",
    "\n",
    "module load bismark\n",
    "\n",
    "FILES=`find ${IN}/ -type f -name \"*.bam\"`\n",
    "#echo ${FILES}\n",
    "\n",
    "for file in $FILES\n",
    "    do\n",
    "        COMMAND=\"bismark_methylation_extractor \\\n",
    "                ${file} \\\n",
    "                -s \\\n",
    "                --merge_non_CpG \\\n",
    "                --cytosine_report \\\n",
    "                --scaffolds \\\n",
    "                --genome_folder ${GENOME} \\\n",
    "                --multicore 6 \\\n",
    "                -o ${OUT}\"\n",
    "      echo $COMMAND\n",
    "\n",
    "done | abatch -j ${OUT}/meth-extract --time 5-01:00:00 --mem 4G -c 1| sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a79f72c3-4b64-4e6a-99b9-5850e8e030fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /workspace/cfngle/raw-data/JM\n",
    "mkdir -p 007b.extraction_nonzero"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "bd0d2f6f-de65-4f03-a15b-ff4cf2fccaa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp 007.meth_extraction/*CpG_report.txt 007b.extraction_nonzero"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d0f06320-01ee-467d-9a82-5602a6fc2168",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /workspace/cfngle/raw-data/JM/007b.extraction_nonzero"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a6faa41b-2d40-4b6f-bd31-27f8fe353734",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i in *CpG_report.txt; do awk '{if ($4 > 0) print}' $i > $i.nonzero.txt; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67750aa2-a035-41fe-a473-5eaaa5ecd7de",
   "metadata": {},
   "source": [
    "##### Extracting all methylation sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "934a5a83-a227-4fc9-a6eb-d392d780dc21",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "awk: fatal: cannot open file `JM_001_trimmed_bismark_bt2.CpG_report.txt.nonzero-checkpoint.txt' for reading (No such file or directory)\n"
     ]
    },
    {
     "ename": "",
     "evalue": "123",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "find . -name \"*.txt\" | xargs -n 1 bash extraction_conversion.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f7068525-3f61-4fb0-8a1f-2a21c1698bc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=1708648365.609395-260321;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.11\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mMultiQC Version v1.20 now available!\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report title: MethylationExtraction_QC\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /workspace/cfndxa/Paua_clock/01_pilot_run/010.extraction\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m433/433\u001b[0m  0m[0m  \n",
      "\u001b[?25h\u001b[34m|           bismark\u001b[0m | Found 24 methextract reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : 010.extraction/MethylationExtraction_QC_multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : 010.extraction/MethylationExtraction_QC_multiqc_report_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "multiqc 010.extraction/ -i MethylationExtraction_QC -o 010.extraction/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7020327-5aba-421e-9fcc-ceddac46cd7f",
   "metadata": {},
   "source": [
    "### Methyl extraction (LOCAL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "366f754b-e37b-48b3-aa23-bd4a6181b268",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bismark_methylation_extractor cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam -s --merge_non_CpG --cytosine_report --scaffolds --genome_folder /workspace/cfngle/raw-data/JM/rgenome --multicore 6 -o /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local\n"
     ]
    }
   ],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "IN=\"${PROJECT}/004.alignments_BM_local\"\n",
    "OUT=\"${PROJECT}/007.meth_extraction_BM_local\"\n",
    "GENOME=\"/workspace/cfngle/raw-data/JM/rgenome\"\n",
    "\n",
    "module load bismark\n",
    "\n",
    "#FILES=`find ${IN}/ -type f -name \"*.bam\"`\n",
    "FILES=\"cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam\"\n",
    "\n",
    "#echo ${FILES}\n",
    "\n",
    "for file in $FILES\n",
    "    do\n",
    "        COMMAND=\"bismark_methylation_extractor \\\n",
    "                ${file} \\\n",
    "                -s \\\n",
    "                --merge_non_CpG \\\n",
    "                --cytosine_report \\\n",
    "                --scaffolds \\\n",
    "                --genome_folder ${GENOME} \\\n",
    "                --multicore 6 \\\n",
    "                -o ${OUT}\"\n",
    "      echo $COMMAND\n",
    "\n",
    "done #| abatch -j ${OUT}/meth-extract --time 23:59:00 --mem 10G -c 1 | sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c9750de8-acce-491d-87f5-27b0b1e11cf2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " *** Bismark methylation extractor version v0.23.0 ***\n",
      "\n",
      "Output will be written into the directory: /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/\n",
      "Setting the option '--bedGraph' since this is required for the genome-wide cytosine report\n",
      "\n",
      "Summarising Bismark methylation extractor parameters:\n",
      "===============================================================\n",
      "Bismark single-end SAM format specified (default)\n",
      "Number of cores to be used: 6\n",
      "Merge CHG and CHH context to non-CpG context specified\n",
      "Output path specified as: /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/\n",
      "\n",
      "\n",
      "Summarising bedGraph parameters:\n",
      "===============================================================\n",
      "Generating additional output in bedGraph and coverage format\n",
      "bedGraph format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage>\n",
      "coverage format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage> <count methylated> <count non-methylated>\n",
      "\n",
      "Using a cutoff of 1 read(s) to report cytosine positions\n",
      "Reporting and sorting cytosine methylation information in CpG context only (default)\n",
      "The bedGraph UNIX sort command will use the following memory setting:\t'2G'. Temporary directory used for sorting is the output directory\n",
      "\n",
      "\n",
      "Summarising genome-wide cytosine methylation report parameters:\n",
      "===============================================================\n",
      "Generating comprehensive genome-wide cytosine report\n",
      "(output format: <Chromosome> <Position> <Strand> <count methylated> <count non-methylated>  <C-context>  <trinucleotide context> )\n",
      "Reporting cytosine methylation in CpG context only (default)\n",
      "Using 1-based genomic coordinates (default)\n",
      "Genome folder was specified as /workspace/cfngle/raw-data/JM/rgenome/\n",
      "\n",
      "Checking file >>/workspace/cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam<< for signs of file truncation...\n",
      "\n",
      "Writing result file containing methylation information for C in CpG context from the original top strand to /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_OT_JM_036_trimmed_bismark_bt2.txt\n",
      "Writing result file containing methylation information for C in CpG context from the complementary to original top strand to /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_CTOT_JM_036_trimmed_bismark_bt2.txt\n",
      "Writing result file containing methylation information for C in CpG context from the complementary to original bottom strand to /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_CTOB_JM_036_trimmed_bismark_bt2.txt\n",
      "Writing result file containing methylation information for C in CpG context from the original bottom strand to /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_OB_JM_036_trimmed_bismark_bt2.txt\n",
      "\n",
      "Writing result file containing methylation information for C in any other context from the original top strand to /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_OT_JM_036_trimmed_bismark_bt2.txt\n",
      "Writing result file containing methylation information for C in any other context from the complementary to original top strand to /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_CTOT_JM_036_trimmed_bismark_bt2.txt\n",
      "Writing result file containing methylation information for C in any other context from the complementary to original bottom strand to /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_CTOB_JM_036_trimmed_bismark_bt2.txt\n",
      "Writing result file containing methylation information for C in any other context from the original bottom strand to /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_OB_JM_036_trimmed_bismark_bt2.bam.txt\n",
      "\n",
      "\n",
      "Now reading in Bismark result file /workspace/cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file /workspace/cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file /workspace/cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file /workspace/cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam\n",
      "\n",
      "Now reading in Bismark result file /workspace/cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam\n",
      "\n",
      "Now reading in Bismark result file /workspace/cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam\n",
      "\n",
      "skipping SAM header line:\t@HD\tVN:1.0\tSO:unsorted\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019859.2\tLN:37713152\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019860.2\tLN:25379070\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019861.2\tLN:38248663\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019862.2\tLN:32868862\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019863.2\tLN:33205099\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019864.2\tLN:32246747\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019865.2\tLN:34573382\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019866.2\tLN:26239357\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019867.2\tLN:33399407\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019868.2\tLN:31218526\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019869.2\tLN:28210532\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019870.2\tLN:30543476\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019871.2\tLN:33825776\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019872.2\tLN:30598983\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019873.2\tLN:30476034\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019874.2\tLN:32958677\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019875.2\tLN:31792230\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019876.2\tLN:30918796\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019877.2\tLN:25472880\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019878.2\tLN:25942153\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019879.2\tLN:31148813\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019880.2\tLN:28976614\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019881.2\tLN:24400806\n",
      "skipping SAM header line:\t@SQ\tSN:NC_019882.2\tLN:23682337\n",
      "skipping SAM header line:\t@SQ\tSN:NC_004387.1\tLN:16714\n",
      "skipping SAM header line:\t@PG\tID:Bismark\tVN:v0.23.0\tCL:\"bismark --genome /workspace/cfngle/raw-data/JM/rgenome /workspace/cfngle/raw-data/JM/002.trim_galore/JM_036_trimmed.fq --o /workspace/cfngle/raw-data/JM/004.alignments_BS_local/JM_036_Alignments --multicore 5 --non_directional --local\"\n",
      "skipping SAM header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.10\tCL:/usr/bin/samtools view -bSh -\n",
      "skipping SAM header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.10\tCL:/usr/bin/samtools view -h /workspace/cfngle/raw-data/JM/004.alignments_BS_local/JM_036_Alignments/JM_036_trimmed.fq.temp.1_bismark_bt2.bam\n",
      "skipping SAM header line:\t@PG\tID:samtools.2\tPN:samtools\tPP:samtools.1\tVN:1.10\tCL:/usr/bin/samtools view -bSh -\n",
      "skipping SAM header line:\t@PG\tID:samtools.3\tPN:samtools\tPP:samtools.2\tVN:1.10\tCL:/usr/bin/samtools view -h /workspace/cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Now waiting for all child processes to complete\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "\n",
      "Merging individual splitting reports into overall report: '/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt'\n",
      "Merging from these individual files:\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.1\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.2\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.3\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.4\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.5\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.6\n",
      "\n",
      "\n",
      "Processed 5255814 lines in total\n",
      "Total number of methylation call strings processed: 5255814\n",
      "\n",
      "Final Cytosine Methylation Report\n",
      "=================================\n",
      "Total number of C's analysed:\t74159684\n",
      "\n",
      "Total methylated C's in CpG context:\t6481638\n",
      "Total methylated C's in CHG context:\t135681\n",
      "Total methylated C's in CHH context:\t740682\n",
      "\n",
      "Total C to T conversions in CpG context:\t4800346\n",
      "Total C to T conversions in CHG context:\t17507335\n",
      "Total C to T conversions in CHH context:\t44494002\n",
      "\n",
      "C methylated in CpG context:\t57.5%\n",
      "C methylated in non-CpG context:\t1.4%\n",
      "\n",
      "\n",
      "\n",
      "Merging individual M-bias reports into overall M-bias statistics from these 6 individual files:\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.1.mbias\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.2.mbias\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.3.mbias\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.4.mbias\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.5.mbias\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2_splitting_report.txt.6.mbias\n",
      "\n",
      "Determining maximum read length for M-Bias plot\n",
      "Maximum read length of Read 1: 70\n",
      "\n",
      "Determining maximum read length for M-Bias plot\n",
      "Maximum read length of Read 1: 70\n",
      "\n",
      "Deleting unused files ...\n",
      "\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_OT_JM_036_trimmed_bismark_bt2.txt contains data ->\tkept\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_CTOT_JM_036_trimmed_bismark_bt2.txt contains data ->\tkept\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_CTOB_JM_036_trimmed_bismark_bt2.txt contains data ->\tkept\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_OB_JM_036_trimmed_bismark_bt2.txt contains data ->\tkept\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_OT_JM_036_trimmed_bismark_bt2.txt contains data ->\tkept\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_CTOT_JM_036_trimmed_bismark_bt2.txt contains data ->\tkept\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_CTOB_JM_036_trimmed_bismark_bt2.txt contains data ->\tkept\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_OB_JM_036_trimmed_bismark_bt2.bam.txt contains data ->\tkept\n",
      "\n",
      "\n",
      "Using these input files: /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_OT_JM_036_trimmed_bismark_bt2.txt /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_CTOT_JM_036_trimmed_bismark_bt2.txt /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_CTOB_JM_036_trimmed_bismark_bt2.txt /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_OB_JM_036_trimmed_bismark_bt2.txt /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_OT_JM_036_trimmed_bismark_bt2.txt /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_CTOT_JM_036_trimmed_bismark_bt2.txt /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_CTOB_JM_036_trimmed_bismark_bt2.txt /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/Non_CpG_OB_JM_036_trimmed_bismark_bt2.bam.txt\n",
      "\n",
      "Summary of parameters for bismark2bedGraph conversion:\n",
      "======================================================\n",
      "bedGraph output:\t\tJM_036_trimmed_bismark_bt2.bedGraph.gz\n",
      "output directory:\t\t>/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/<\n",
      "remove whitespaces:\t\tno\n",
      "CX context:\t\t\tno (CpG context only, default)\n",
      "No-header selected:\t\tno\n",
      "Sorting method:\t\t\tUnix sort-based (smaller memory footprint, but slower)\n",
      "Sort buffer size:\t\t2G\n",
      "Coverage threshold:\t\t1\n",
      "=============================================================================\n",
      "Methylation information will now be written into a bedGraph and coverage file\n",
      "=============================================================================\n",
      "\n",
      "Using the following files as Input:\n",
      "/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_OT_JM_036_trimmed_bismark_bt2.txt\t/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_CTOT_JM_036_trimmed_bismark_bt2.txt\t/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_CTOB_JM_036_trimmed_bismark_bt2.txt\t/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/CpG_OB_JM_036_trimmed_bismark_bt2.txt\n",
      "\n",
      "Writing bedGraph to file: JM_036_trimmed_bismark_bt2.bedGraph.gz\n",
      "Also writing out a coverage file including counts methylated and unmethylated residues to file: JM_036_trimmed_bismark_bt2.bismark.cov.gz\n",
      "\n",
      "Changed directory to /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/\n",
      "The genome of interest was specified to contain gazillions of chromosomes or scaffolds. Merging all input files and sorting everything in memory instead of writing out individual chromosome files...\n",
      "Writing all merged methylation calls to temp file JM_036_trimmed_bismark_bt2.bedGraph.gz.methylation_calls.merged\n",
      "\n",
      "Finished writing methylation calls from CpG_OT_JM_036_trimmed_bismark_bt2.txt to merged temp file\n",
      "Finished writing methylation calls from CpG_CTOT_JM_036_trimmed_bismark_bt2.txt to merged temp file\n",
      "Finished writing methylation calls from CpG_CTOB_JM_036_trimmed_bismark_bt2.txt to merged temp file\n",
      "Finished writing methylation calls from CpG_OB_JM_036_trimmed_bismark_bt2.txt to merged temp file\n",
      "Sorting input file JM_036_trimmed_bismark_bt2.bedGraph.gz.methylation_calls.merged by positions (using -S of 2G)\n",
      "Successfully deleted the temporary input file JM_036_trimmed_bismark_bt2.bedGraph.gz.methylation_calls.merged\n",
      "\n",
      "Finished BedGraph conversion ...\n",
      "\n",
      "Output will be written into the directory: /workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/\n",
      "Summary of parameters for genome-wide cytosine report:\n",
      "==============================================================================\n",
      "Coverage infile:\t\t\t\tJM_036_trimmed_bismark_bt2.bismark.cov.gz\n",
      "Output directory:\t\t\t\t>/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/<\n",
      "Parent directory:\t\t\t\t>/workspace/cfngle/raw-data/JM/007.meth_extraction_BM_local/<\n",
      "Genome directory:\t\t\t\t>/workspace/cfngle/raw-data/JM/rgenome/<\n",
      "Coverage threshold\t\t\t\t0 (all positions will be reported, default)\n",
      "CX context:\t\t\t\t\tno (CpG context only, default)\n",
      "Genome coordinates used:\t\t\t1-based (default)\n",
      "GZIP compression:\t\t\t\tno\n",
      "Split by chromosome:\t\t\t\tno\n",
      "\n",
      "\n",
      "Now reading in and storing sequence information of the genome specified in: /workspace/cfngle/raw-data/JM/rgenome/\n",
      "\n",
      "Couldn't find files ending in .fa, trying .fa.gz instead\n",
      "chr NC_019859.2 (37713152 bp)\n",
      "chr NC_019860.2 (25379070 bp)\n",
      "chr NC_019861.2 (38248663 bp)\n",
      "chr NC_019862.2 (32868862 bp)\n",
      "chr NC_019863.2 (33205099 bp)\n",
      "chr NC_019864.2 (32246747 bp)\n",
      "chr NC_019865.2 (34573382 bp)\n",
      "chr NC_019866.2 (26239357 bp)\n",
      "chr NC_019867.2 (33399407 bp)\n",
      "chr NC_019868.2 (31218526 bp)\n",
      "chr NC_019869.2 (28210532 bp)\n",
      "chr NC_019870.2 (30543476 bp)\n",
      "chr NC_019871.2 (33825776 bp)\n",
      "chr NC_019872.2 (30598983 bp)\n",
      "chr NC_019873.2 (30476034 bp)\n",
      "chr NC_019874.2 (32958677 bp)\n",
      "chr NC_019875.2 (31792230 bp)\n",
      "chr NC_019876.2 (30918796 bp)\n",
      "chr NC_019877.2 (25472880 bp)\n",
      "chr NC_019878.2 (25942153 bp)\n",
      "chr NC_019879.2 (31148813 bp)\n",
      "chr NC_019880.2 (28976614 bp)\n",
      "chr NC_019881.2 (24400806 bp)\n",
      "chr NC_019882.2 (23682337 bp)\n",
      "chr NC_004387.1 (16714 bp)\n",
      "\n",
      "Stored sequence information of 25 chromosomes/scaffolds in total\n",
      "\n",
      "==============================================================================\n",
      "Methylation information will now be written into a genome-wide cytosine report\n",
      "==============================================================================\n",
      "\n",
      "Adding context-specific methylation summaries\n",
      "\n",
      ">>> Writing genome-wide cytosine report to: JM_036_trimmed_bismark_bt2.CpG_report.txt <<<\n",
      "\n",
      ">>> Writing all cytosine context summary file to: JM_036_trimmed_bismark_bt2.cytosine_context_summary.txt <<<\n",
      "\n",
      "Storing all covered cytosine positions for chromosome: NC_004387.1\n",
      "Writing cytosine report for chromosome NC_004387.1 (stored 212 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019859.2 (stored 76806 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019860.2 (stored 63713 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019861.2 (stored 62867 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019862.2 (stored 55535 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019863.2 (stored 55690 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019864.2 (stored 54900 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019865.2 (stored 57554 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019866.2 (stored 58411 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019867.2 (stored 57391 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019868.2 (stored 50348 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019869.2 (stored 54578 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019870.2 (stored 63179 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019871.2 (stored 60561 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019872.2 (stored 51507 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019873.2 (stored 50313 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019874.2 (stored 57173 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019875.2 (stored 58525 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019876.2 (stored 63440 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019877.2 (stored 46607 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019878.2 (stored 44831 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019879.2 (stored 52465 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019880.2 (stored 63234 different covered positions)\n",
      "Writing cytosine report for chromosome NC_019881.2 (stored 58193 different covered positions)\n",
      "Writing cytosine report for last chromosome NC_019882.2 (stored 44357 different covered positions)\n",
      "Finished writing out cytosine report for covered chromosomes (processed 25 chromosomes/scaffolds in total)\n",
      "\n",
      "Now processing chromosomes that were not covered by any methylation calls in the coverage file...\n",
      "All chromosomes in the genome were covered by at least some reads. coverage2cytosine processing complete.\n",
      "\n",
      "\n",
      "\n",
      "Finished generating genome-wide cytosine report\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## verifying \n",
    "\n",
    "PROJECT=\"/workspace/cfngle/raw-data/JM\"\n",
    "IN=\"${PROJECT}/004.alignments_BM_local\"\n",
    "OUT=\"${PROJECT}/007.meth_extraction_BM_local\"\n",
    "GENOME=\"/workspace/cfngle/raw-data/JM/rgenome\"\n",
    "\n",
    "module load bismark\n",
    "\n",
    "#FILES=`find ${IN}/ -type f -name \"*.bam\"`\n",
    "FILES=\"/workspace/cfngle/raw-data/JM/004.alignments_BM_local/JM_036_Alignments/JM_036_trimmed_bismark_bt2.bam\"\n",
    "\n",
    "#echo ${FILES}\n",
    "\n",
    "bismark_methylation_extractor ${FILES} -s --merge_non_CpG --cytosine_report --scaffolds --genome_folder ${GENOME} --multicore 6 -o ${OUT}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bc70ae95-e782-43fa-a3b5-162f4bc153ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /workspace/cfngle/raw-data/JM\n",
    "mkdir -p 007b.extraction_nonzero_BM_local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "c500480d-b0b0-4225-ad79-46c45f70edf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp 007.meth_extraction_BM_local/*CpG_report.txt 007b.extraction_nonzero_BM_local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1fe08a6f-be76-4be0-9fe1-83e3fd389fbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## cp 007.meth_extraction_BM_local/JM_036_trimmed_bismark_bt2.CpG_report.txt 007b.extraction_nonzero_BM_local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cab380b6-fd58-4436-8c0f-919cb840f0b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /workspace/cfngle/raw-data/JM/007b.extraction_nonzero_BM_local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "c229e513-2d97-437c-a34c-7374f8a262e4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i in *CpG_report.txt; do awk '{if ($4 > 0) print}' $i > $i.nonzero.txt; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ce7edb5a-1bc9-4d7a-bf5d-2c0726a6e9ea",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## for i in JM_036_trimmed_bismark_bt2.CpG_report.txt; do awk '{if ($4 > 0) print}' $i > $i.nonzero.txt; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eaddbe1-bc50-4301-89a0-ef628d3b78d8",
   "metadata": {},
   "source": [
    "##### Extracting all methylation sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "956b482e-62ae-4e68-be42-6d2eed071bd2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "find . -name \"*nonzero.txt\" | xargs -n 1 bash extraction_conversion.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "82da6888-28e6-44c9-a881-7a162410d87d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "find . -name \"JM_036_trimmed_bismark_bt2.CpG_report.txt.nonzero.txt\" | xargs -n 1 bash extraction_conversion.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b3bb9d3d-89bc-4fa5-8eaf-b0b45b83225b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mv *MK.txt /workspace/cfngle/raw-data/JM/008.methylkit_BM_local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5bb52aa-11a0-4802-8ffb-c45a7926b618",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
