{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2685f3a6-cf6b-4c7d-bac3-37ca3d5157d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently Loaded Modulefiles:\n",
      " 1) \u001b[46mpowerPlant/core\u001b[0m    5) \u001b[46mSlurm/21.08.8-2\u001b[0m      9) bowtie2/2.3.4.3  \n",
      " 2) texlive/20230914   6) perlbrew/0.76       10) minimap2/2.22    \n",
      " 3) pandoc/1.19.2      7) perl/5.36.0         11) cd-hit/4.6.1     \n",
      " 4) git/2.21.0         8) slurm-utils/latest  12) samtools/1.16    \n",
      "\n",
      "Key:\n",
      "\u001b[46msticky\u001b[0m  \n"
     ]
    }
   ],
   "source": [
    "# Alignments for various species to reference genomes\n",
    "\n",
    "# loading modules\n",
    "module load bowtie2\n",
    "module load minimap2 \n",
    "module load cd-hit\n",
    "module load samtools\n",
    "\n",
    "# checking active \n",
    "module list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3eec4422-fd71-4360-90d1-1fe9b4aaddfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# setting up variables \n",
    "# paths\n",
    "path_raw=/workspace/cfngle/raw-data/\n",
    "path_results=/workspace/cfngle/results-data/\n",
    "path_sequences=/workspace/cfngle/results-data/sequences/\n",
    "\n",
    "# filenames\n",
    "seq_100bp=(\"AC_CpG_100bp.fasta\" \"AS_CpG_100bp.fasta\" \"EH_CpG_100bp.fasta\")\n",
    "\n",
    "seq_200bp=(\"AC_CpG_200bp.fasta\" \"AS_CpG_200bp.fasta\" \"EH_CpG_200bp.fasta\")\n",
    "\n",
    "seq_500bp=(\"AC_CpG_500bp.fasta\" \"AS_CpG_500bp.fasta\" \"EH_CpG_500bp.fasta\")\n",
    "\n",
    "seq_1000bp=(\"AC_CpG_1000bp.fasta\" \"AS_CpG_1000bp.fasta\" \"EH_CpG_1000bp.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd4afcc3-a0e0-4fae-a616-e9b8dd291e89",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### CD-HIT ####\n",
    "# clustering sequences with cd-hit\n",
    "cdhit_opt=\"-p 1 -c 0.75 -n 4 -T 0 -M 0\"\n",
    "cdhit_nmsfx=\"_75\"\n",
    "\n",
    "# AC\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[0]} -i2 ${path_results}${seq_200bp[1]} -o ${path_results}cd-hit/AC_AS_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[0]} -i2 ${path_results}${seq_200bp[2]} -o ${path_results}cd-hit/AC_EH_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "# AS\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[1]} -i2 ${path_results}${seq_200bp[2]} -o ${path_results}cd-hit/AS_EH_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[1]} -i2 ${path_results}${seq_200bp[0]} -o ${path_results}cd-hit/AS_AC_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "# EH\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[2]} -i2 ${path_results}${seq_200bp[0]} -o ${path_results}cd-hit/EH_AC_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "cd-hit-est-2d -i ${path_results}${seq_200bp[2]} -i2 ${path_results}${seq_200bp[1]} -o ${path_results}cd-hit/EH_AS_cd-hit${cdhit_nmsfx}.fasta $cdhit_opt\n",
    "\n",
    "\n",
    "# loop test\n",
    ": << 'COMMENT'\n",
    "for i in $seq_200bp; do\n",
    "    cd-hit-est-2d -i ${path_results}$i -i2 ${path_results}AC -o ${path_results}cd-hit/${i}cd-hit_AC.fasta -c 0.9 -p 1\n",
    "done\n",
    "COMMENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "2bf8e06c-ab45-472d-b55d-0384f38d0bcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13900 reads; of these:\n",
      "  13900 (100.00%) were unpaired; of these:\n",
      "    12449 (89.56%) aligned 0 times\n",
      "    487 (3.50%) aligned exactly 1 time\n",
      "    964 (6.94%) aligned >1 times\n",
      "10.44% overall alignment rate\n",
      "27560 reads; of these:\n",
      "  27560 (100.00%) were unpaired; of these:\n",
      "    26269 (95.32%) aligned 0 times\n",
      "    643 (2.33%) aligned exactly 1 time\n",
      "    648 (2.35%) aligned >1 times\n",
      "4.68% overall alignment rate\n",
      "20018 reads; of these:\n",
      "  20018 (100.00%) were unpaired; of these:\n",
      "    16588 (82.87%) aligned 0 times\n",
      "    792 (3.96%) aligned exactly 1 time\n",
      "    2638 (13.18%) aligned >1 times\n",
      "17.13% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "#### BOWTIE2 ####\n",
    "\n",
    "## modules\n",
    "module load bowtie\n",
    "\n",
    "## setting up variables \n",
    "# paths\n",
    "path_raw=/workspace/cfngle/raw-data/\n",
    "path_results=/workspace/cfngle/results-data/bowtie2/\n",
    "path_sequences=/workspace/cfngle/results-data/sequences/\n",
    "\n",
    "# filenames\n",
    "seq_100bp=(\"AC_CpG_100bp\" \"AS_CpG_100bp\" \"EH_CpG_100bp\")\n",
    "seq_200bp=(\"AC_CpG_200bp\" \"AS_CpG_200bp\" \"EH_CpG_200bp\")\n",
    "seq_500bp=(\"AC_CpG_500bp\" \"AS_CpG_500bp\" \"EH_CpG_500bp\")\n",
    "seq_1000bp=(\"AC_CpG_1000bp\" \"AS_CpG_1000bp\" \"EH_CpG_1000bp\")\n",
    "\n",
    "seq_all=($seq_100bp, $seq_200bp, $seq_500bp, $seq_1000bp,)\n",
    "\n",
    "# file ending\n",
    "suffix=\".fasta\"\n",
    "\n",
    "# arguments\n",
    "bowtie2_args=\"--very-sensitive --local -p 8\"\n",
    "\n",
    "\n",
    "for (( i=0; i<${#seq_100bp[@]}; i++ )); \n",
    "do \n",
    "    bowtie2 $bowtie2_args -x ${path_raw}ZF/rgenome/bowtie2-index/ZF_bowtie2 -f -U ${path_sequences}${seq_100bp[$i]}$suffix -S ${path_results}ZF_${seq_100bp[$i]}_bt2_.sam -N 1\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a6903d20-9dbf-45ef-8555-3bba82d40e5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "#### MINIMAP2 ####\n",
    "\n",
    "#ZF\n",
    "#AC\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_200bp[0]} > ${path_results}minimap2/ZF_AC_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_100bp[0]} > ${path_results}minimap2/ZF_AC_100_minimap.sam\n",
    "#AS\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_200bp[1]} > ${path_results}minimap2/ZF_AS_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_100bp[1]} > ${path_results}minimap2/ZF_AS_100_minimap.sam\n",
    "#EH\n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_200bp[2]} > ${path_results}minimap2/ZF_EH_200_minimap.sam \n",
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_results}${seq_100bp[2]} > ${path_results}minimap2/ZF_EH_100_minimap.sam \n",
    "\n",
    "path_rgenome=\"AC/GCF_902167405.1_gadMor3.0_genomic.fasta\"\n",
    "#AC\n",
    "#AC\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[0]} > ${path_results}minimap2/AC_AC_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[0]} > ${path_results}minimap2/AC_AC_100_minimap.sam\n",
    "#AS\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[1]} > ${path_results}minimap2/AC_AS_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[1]} > ${path_results}minimap2/AC_AS_100_minimap.sam\n",
    "#EH\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[2]} > ${path_results}minimap2/AC_EH_200_minimap.sam \n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[2]} > ${path_results}minimap2/AC_EH_100_minimap.sam \n",
    "\n",
    "path_rgenome=\"Chrysophrys_auratus.v.1.0.all.male.map.fasta\"\n",
    "#AS\n",
    "#AC\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[0]} > ${path_results}minimap2/AS_AC_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[0]} > ${path_results}minimap2/AS_AC_100_minimap.sam\n",
    "#AS\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[1]} > ${path_results}minimap2/AS_AS_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[1]} > ${path_results}minimap2/AS_AS_100_minimap.sam\n",
    "#EH\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[2]} > ${path_results}minimap2/AS_EH_200_minimap.sam \n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[2]} > ${path_results}minimap2/AS_EH_100_minimap.sam \n",
    "\n",
    "path_rgenome=\"fMerMel2.1_cnag1.scaffolds.fa\"\n",
    "#EH\\\n",
    "#AC\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[0]} > ${path_results}minimap2/EH_AC_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[0]} > ${path_results}minimap2/EH_AC_100_minimap.sam\n",
    "#AS\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[1]} > ${path_results}minimap2/EH_AS_200_minimap.sam\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[1]} > ${path_results}minimap2/EH_AS_100_minimap.sam\n",
    "#EH\n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_200bp[2]} > ${path_results}minimap2/EH_EH_200_minimap.sam \n",
    "minimap2 -ax map-ont ${path_raw}path_rgenome ${path_results}${seq_100bp[2]} > ${path_results}minimap2/EH_EH_100_minimap.sam \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ef724657-b0c3-4a76-b119-afe8bcbef7ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "173\n",
      "415\n",
      "1391\n",
      "2680\n",
      "265\n",
      "497\n",
      "1081\n",
      "1975\n",
      "456\n",
      "926\n",
      "2863\n",
      "5472\n",
      "16349\n",
      "15460\n",
      "14918\n",
      "14480\n",
      "1397\n",
      "2073\n",
      "3218\n",
      "4906\n",
      "10165\n",
      "12835\n",
      "18254\n",
      "25553\n",
      "844\n",
      "1231\n",
      "2149\n",
      "3650\n",
      "28688\n",
      "24797\n",
      "24642\n",
      "24530\n",
      "2261\n",
      "3698\n",
      "5642\n",
      "8460\n",
      "5126\n",
      "7118\n",
      "12052\n",
      "18420\n",
      "1588\n",
      "2508\n",
      "4191\n",
      "6812\n",
      "24022\n",
      "21442\n",
      "20341\n",
      "19491\n"
     ]
    }
   ],
   "source": [
    "## view how many alignments (multiple alignments)\n",
    "\n",
    "##ZF\n",
    "#AC\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AC_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AC_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AC_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AC_1000_minimap.sam\n",
    "#AS\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AS_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AS_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AS_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_AS_1000_minimap.sam\n",
    "#EH\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_EH_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_EH_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_EH_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/ZF_EH_1000_minimap.sam\n",
    "\n",
    "##AC\n",
    "#AC\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AC_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AC_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AC_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AC_1000_minimap.sam\n",
    "#AS\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AS_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AS_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AS_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_AS_1000_minimap.sam\n",
    "#EH\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_EH_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_EH_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_EH_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AC_EH_1000_minimap.sam\n",
    "\n",
    "##AS\n",
    "#AC\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AC_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AC_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AC_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AC_1000_minimap.sam\n",
    "#AS\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AS_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AS_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AS_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_AS_1000_minimap.sam\n",
    "#EH\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_EH_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_EH_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_EH_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/AS_EH_1000_minimap.sam\n",
    "\n",
    "##EH\n",
    "#AC\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AC_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AC_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AC_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AC_1000_minimap.sam\n",
    "#AS\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AS_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AS_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AS_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_AS_1000_minimap.sam\n",
    "#EH\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_EH_100_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_EH_200_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_EH_500_minimap.sam\n",
    "samtools view -c -F 4 /workspace/cfngle/results-data/minimap2/EH_EH_1000_minimap.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "05689c53-68b5-42de-bd55-bf1a229e93b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "107\n",
      "207\n",
      "515\n",
      "887\n",
      "158\n",
      "280\n",
      "543\n",
      "896\n",
      "268\n",
      "430\n",
      "970\n",
      "1662\n",
      "13879\n",
      "13755\n",
      "13720\n",
      "13641\n",
      "802\n",
      "1165\n",
      "1861\n",
      "2809\n",
      "6253\n",
      "7767\n",
      "10012\n",
      "12323\n",
      "514\n",
      "768\n",
      "1277\n",
      "2051\n",
      "27538\n",
      "24535\n",
      "24489\n",
      "24410\n",
      "1252\n",
      "1768\n",
      "2666\n",
      "3912\n",
      "3888\n",
      "5192\n",
      "7301\n",
      "9543\n",
      "958\n",
      "1467\n",
      "2493\n",
      "3817\n",
      "19904\n",
      "18329\n",
      "18192\n",
      "18049\n"
     ]
    }
   ],
   "source": [
    "# view single alignments\n",
    "\n",
    "##ZF\n",
    "#AC\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AC_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AC_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AC_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AC_1000_minimap.sam | wc -l\n",
    "#AS\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AS_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AS_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AS_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_AS_1000_minimap.sam | wc -l\n",
    "#EH\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_EH_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_EH_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_EH_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_EH_1000_minimap.sam | wc -l\n",
    "\n",
    "##AC\n",
    "#AC\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AC_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AC_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AC_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AC_1000_minimap.sam | wc -l\n",
    "#AS\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AS_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AS_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AS_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_AS_1000_minimap.sam | wc -l\n",
    "#EH\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_EH_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_EH_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_EH_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AC_EH_1000_minimap.sam | wc -l\n",
    "\n",
    "##AS\n",
    "#AC\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AC_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AC_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AC_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AC_1000_minimap.sam | wc -l\n",
    "#AS\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AS_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AS_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AS_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_AS_1000_minimap.sam | wc -l\n",
    "#EH\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_EH_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_EH_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_EH_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/AS_EH_1000_minimap.sam | wc -l\n",
    "\n",
    "##EH\n",
    "#AC\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AC_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AC_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AC_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AC_1000_minimap.sam | wc -l\n",
    "#AS\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AS_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AS_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AS_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_AS_1000_minimap.sam | wc -l\n",
    "#EH\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_EH_100_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_EH_200_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_EH_500_minimap.sam | wc -l\n",
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/EH_EH_1000_minimap.sam | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3eafa8c1-2bc0-4d30-ad71-4e09ca7bbb37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::32.235*1.63] collected minimizers\n",
      "[M::mm_idx_gen::40.264*1.89] sorted minimizers\n",
      "[M::main::40.266*1.89] loaded/built the index for 1923 target sequence(s)\n",
      "[M::mm_mapopt_update::41.644*1.86] mid_occ = 848\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1923\n",
      "[M::mm_idx_stat::42.496*1.84] distinct minimizers: 66057654 (43.51% are singletons); average occurrences: 4.867; average spacing: 5.223; total length: 1679203469\n",
      "[M::worker_pipeline::66.027*2.21] mapped 16751 sequences\n",
      "[M::main] Version: 2.22-r1101\n",
      "[M::main] CMD: minimap2 -ax map-ont /workspace/cfngle/raw-data/ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna /workspace/cfngle/results-data/sequences/EH_CpG_5000bp.fasta\n",
      "[M::main] Real time: 66.285 sec; CPU: 146.265 sec; Peak RSS: 6.608 GB\n"
     ]
    }
   ],
   "source": [
    "minimap2 -ax map-ont ${path_raw}ZF/rgenome/GCF_000002035.6_GRCz11_genomic.fna ${path_sequences}EH_CpG_5000bp.fasta > ${path_results}minimap2/ZF_EH_5000_minimap.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6a2dff7d-0080-4bfd-909b-726e2f3eefd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5557\n"
     ]
    }
   ],
   "source": [
    "samtools view -F 0x904 /workspace/cfngle/results-data/minimap2/ZF_EH_5000_minimap.sam | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f18465a1-e736-45c5-8a1a-9f8c8355800b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### MAFFT ####\n",
    "\n",
    "## modules\n",
    "module load mafft\n",
    "\n",
    "## setting up variables \n",
    "\n",
    "# paths\n",
    "path_raw=/workspace/cfngle/raw-data/\n",
    "path_results=/workspace/cfngle/results-data/mafft/\n",
    "path_sequences=/workspace/cfngle/results-data/sequences/\n",
    "\n",
    "# arguments\n",
    "mafft_args=\"--auto --thread -1\"\n",
    "\n",
    "mafft $mafft_args ${path_sequences}AC_AS_EH_CpG_100bp.fasta > ${path_results}AC_AS_EH_CpG_100bp.fasta \n",
    "mafft $mafft_args ${path_sequences}AC_AS_EH_CpG_200bp.fasta > ${path_results}AC_AS_EH_CpG_200bp.fasta \n",
    "mafft $mafft_args ${path_sequences}AC_AS_EH_CpG_500bp.fasta > ${path_results}AC_AS_EH_CpG_500bp.fasta \n",
    "mafft $mafft_args ${path_sequences}AC_AS_EH_CpG_1000bp.fasta > ${path_results}AC_AS_EH_CpG_1000bp.fasta "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
